---
title: "LI paper scripts"
author: 
  - "Mendes, GA"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    code_download: yes
    df_print: paged
    keep_md: yes
    theme: flatly
    toc: true
    toc_depth: 5
    toc_float: true
  pdf_document: 
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}

# load project environment and Rmd necessary libraries
# load(file = "/home/gabriel/projetos/LI_paper/Environments/...RData")

```


**This pipeline integrates public tools available for metagenomic analyses. To share or reproduce this content, please require authors' consent.**  
**Contact:** gabrielmendesbrt@gmail.com 

# Bioinformatics

## Load  R libs

```{r, eval=FALSE,echo=FALSE}
# Loading libraries ----
{ 
  library("Biostrings")
  library("DECIPHER")
  library("phyloseq")
  library("vegan")
  library("tidyverse")
  library("dplyr")
  library("readxl")
  library("ggpubr")
  library("ggplot2")
  library("ggsignif")
  }
```

## Create project folders

```{r, eval=FALSE,echo=FALSE}
# Creating and loading paths ----
{
  prjct_path <- "/home/gabriel/projetos/LI_paper"
  prjct_radical <- "eDNA_Lagoa-dos-Ingleses"
  scripts_path <- paste0(prjct_path,"/scripts")
  results_path <- paste0(prjct_path,"/results")
  figs_path <- paste0(results_path,"/figures")
  tbl_path <- paste0(results_path,"/tables")
  env_path <- paste0(prjct_path,"/environment")
  paths <- c(scripts_path, results_path, figs_path, tbl_path, env_path)
}

for (dir in paths) {
  if (dir.exists(dir)) {
    print(paste("The directory", dir, "already exists!"))
  } else {
    print(paste("Making", dir, "!" ))
    dir.create(dir) 
  }
}
```

## Data acquiring

```{r, eval=FALSE,echo=FALSE}
pre_raw_results_tbl <- read_excel("/home/gabriel/projetos/lagoa_ingleses/tabelas/curated/curated-Complete_analysis_results-2024-01-10.xlsx") %>% tibble()
curated_ids_tbl <- read_excel("/home/gabriel/projetos/lagoa_ingleses/tabelas/curated/curated_lagoa_ingleses-ASVs_x_amostras-2024-01-09.xlsx") %>% tibble()
blast_tax <- read.csv("/home/gabriel/projetos/lagoa_ingleses/tabelas/raw/tax_blast.csv", sep = ",", check.names = FALSE)
```

## Data merging and transformation 

```{r, eval=FALSE,echo=FALSE}
## Selecting colunms ----

# Curated IDs

curated_ids_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()

curated_ids_tbl <-
  curated_ids_tbl %>%
  select(c( "ASV header", "ASV (Sequence)", "Curated ID", "Final ID (BLASTn)", "BLASTn pseudo-score", 
            "Class (BLASTn)", "Curated Order (BLASTn)")) %>%
  unique()

# View(curated_ids_tbl)

# Complete analysis

# pre_raw_results_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()

pre_raw_results_tbl <-
  pre_raw_results_tbl %>%
  select(c("Sample", "Unique_File_name", "OTU", "Abundance", "Obs. Curadoria", "Possible contamination", 
           "Read origin", "Primer expected length", "ASV Size (pb)", "ASV header")) %>%
  mutate(Sample = str_replace(Sample, "__", "_"))

# View(pre_raw_results_tbl)

## Integrating curated identifications and reordering ----

# pre_raw_results_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()

raw_results_tbl <- pre_raw_results_tbl %>%
  left_join(curated_ids_tbl,
            by = c("ASV header")) %>%
  select(c("ASV header", #oriundas da curated_ids_tbl
           "ASV (Sequence)",
           "Curated ID",
           "Final ID (BLASTn)",
           "BLASTn pseudo-score",
           "Class (BLASTn)",
           "Curated Order (BLASTn)",
           "Sample", #oriundas da pre_raw_samples_tbl
           "Unique_File_name",
           "OTU",
           "Abundance",
           "Obs. Curadoria",
           "Read origin",
           "Primer expected length",
           "ASV Size (pb)",
           "ASV header"
  ))

# View(raw_results_tbl)

## Adding BLASTn's complete taxonomy ----

blast_tax <- blast_tax[2:85] %>% 
  rename(`ASV (Sequence)` = "ASV")

blast_tax_less <- blast_tax %>%
  left_join(raw_results_tbl,
           by = "ASV (Sequence)") %>% 
  mutate("Genus (BLASTn)" = str_split_fixed(string = .$`Curated ID`, 
                                           pattern = " ",
                                           n = 2)[,1]) %>% 
  select(c("ASV (Sequence)",
           # "Order (BLASTn)",
           "Family (BLASTn)",
           "Genus (BLASTn)"))
         
# raw_results_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()
# blast_tax_less %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()

raw_results_tbl <- raw_results_tbl %>%
  left_join(blast_tax_less,
            by = "ASV (Sequence)") %>% 
  select(c("ASV header", "ASV (Sequence)", "Curated ID", "Final ID (BLASTn)", "BLASTn pseudo-score", 
           "Class (BLASTn)", "Family (BLASTn)", "Genus (BLASTn)", "Curated Order (BLASTn)", "Sample", 
           "Unique_File_name", "OTU", "Abundance", "Obs. Curadoria", "Read origin", "Primer expected length", 
           "ASV Size (pb)")) %>% 
  unique() %>% 
  rename("Order (BLASTn)" = "Curated Order (BLASTn)")

# View(raw_results_tbl)

## 4o Preliminary filtering ----

# Filtering out of range reads, NA identifications, Mammalian and Aves reads
# with BLASTn pseudo-score lower than 98 and reads from negative controls

pre_filt_tbl <-
  raw_results_tbl %>%
  filter(!`Primer expected length` %in% "out of range") %>% 
  filter(!`Curated ID` == "NA") %>% 
  filter(!(`Class (BLASTn)` %in% c("Mammalia", "Aves") & `BLASTn pseudo-score` < 98)) %>% 
  filter(!(Sample %in% c("EM113_NEGPCR1", "EM135c4c5_NEGPCR1", "EM135c4c5_NEGPCR2", "EM149_NEGPCR1", 
                         "EM149_NEGPCR2", "EM156_NegPCR1", "S724_NEGPCR1", "br_jan_22", "br_nov21"))) %>% 
  # Filter(!(`Possible contamination` %in% c("Possible contamination"))) %>% 
  filter(!(`Obs. Curadoria` %in% c("Possible contamination")))

# View(pre_filt_tbl)

## 5o Adding metadata from sampling 

# pre_filt_tbl$Sample %>% unique() %>% sort() %>% paste0(collapse = '",\n"') %>% cat()
{
  # sample
  Sample <- c("L1_nov_dec_20_mi", "L2_nov20", "L2_dez20", "L1_out21", "L2_out21", "L3_out21", 
              "L4_out21", "L1_nov21", "L2_nov21", "L3_nov21", "L4_nov21", "STX_L1_nov21", 
              "STX_L2_nov21", "STX_L3_nov21", "STX_L4_nov21", "L1_jan22", "L2_jan22", "L3_jan22", 
              "L4_jan22")
  
  # Expedition
  Expedition <- c("Novembro e Dezembro 2020", "Novembro 2020", "Dezembro 2020", "Outubro 2021", 
                  "Outubro 2021", "Outubro 2021", "Outubro 2021", "Novembro 2021", "Novembro 2021", 
                  "Novembro 2021", "Novembro 2021", "Novembro 2021", "Novembro 2021", "Novembro 2021", 
                  "Novembro 2021", "Janeiro 2022", "Janeiro 2022", "Janeiro 2022", "Janeiro 2022")
  
  # Year
  Year <- c("2020", "2020", "2020", "2021", "2021", "2021", "2021", "2021", "2021", "2021", "2021", 
            "2021", "2021", "2021", "2021", "2022", "2022", "2022", "2022")
  
  # Site
  Site <- c("Fundação", "Ponte", "Ponte", "Prainha", "Barragem", "Ponte", "Fundação", "Prainha", 
            "Barragem", "Ponte", "Fundação", "Prainha", "Barragem", "Ponte", "Fundação", "Prainha", 
            "Barragem", "Ponte", "Fundação")
  
  # Filter
  Filter <- c("MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "Sterivex", 
              "Sterivex", "Sterivex", "Sterivex", "MCE", "MCE", "MCE", "MCE")
  
  # Run
  Run <- c("Run2_ago21", "Run4_out21", "Run4_out21", "Run5_dez21", "Run5_dez21", "Run5_dez21", 
           "Run5_dez21", "Run5_dez21", "Run5_dez21", "Run5_dez21", "Run5_dez21", "EM156", "EM156", 
           "EM156", "EM156", "EM156", "EM156", "EM156", "EM156")

  # Level
  Level <- c("Full", "Full", "Full", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", 
             "Low", "Low", "Low", "Full", "Full", "Full", "Full")
}

# As tibble
metadata_tbl <- tibble(Run, Sample, Filter, Site, Expedition, Year, Level)

# Merging metadata with filtered results

# pre_filt_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()
# metadata_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()

# pre_filt_tbl$Sample %>% unique()

curated_full_tbl <- left_join(pre_filt_tbl, metadata_tbl,
                              by = "Sample") %>%
  select(c("ASV header", "ASV (Sequence)", "Curated ID", "Final ID (BLASTn)", "BLASTn pseudo-score", 
           "Class (BLASTn)", "Family (BLASTn)", "Genus (BLASTn)", "Order (BLASTn)", "Sample", 
           "Unique_File_name", "OTU", "Abundance", "Read origin", "Primer expected length", "ASV Size (pb)", 
           "Run", "Sample", "Filter", "Site", "Expedition", "Year", "Level")) 

# View(curated_full_tbl)
```

## Suplementary tables
```{r, eval=FALSE,echo=FALSE}
## Suplementar table A ----

# Creating a table to use to evaluate the unfiltered data

# 1st Longer table
dt_all_resume <- curated_full_tbl %>% 
  group_by(Level, Site, Filter, Year) %>%
  mutate("Abd total" = sum(Abundance)) %>%
  ungroup() %>%
  group_by(`Curated ID`) %>% 
  mutate("Abundancia total" = sum(Abundance)) %>% 
  ungroup %>% 
  group_by(`Final ID (BLASTn)`) %>% 
  mutate("Reads totais" = sum(Abundance)) %>% 
  ungroup() %>% 
  group_by(`Curated ID`, `Final ID (BLASTn)`, `Site`, 
           `Level`, `Filter`, `Year`) %>%
  mutate("RRA no periodo" = (Abundance/`Abd total`)*100) %>% 
  reframe(
    "Total reads" = `Reads totais`,
    "Total abundance" = `Abundancia total`,
    "RRA" = sum(`RRA no periodo`),
    "ASVs" = length(unique(`ASV header`)),
    "OTUs" = length(unique(OTU)),
    "Reads" = sum(Abundance),
    "Class" = `Class (BLASTn)`,
    "Order" = `Order (BLASTn)`,
    "Family" = `Family (BLASTn)`,
    "Genus" = `Genus (BLASTn)`
  ) %>%
  unique() 

# View(dt_all_resume)

# Testing if RRA was calculatad right
dt_all_resume %>%
  group_by( `Site`, `Level`, `Filter`, `Year`) %>%
  summarise(total_RRA = sum(RRA, na.rm = TRUE)) %>%
  ungroup() %>% View()

# Wider table to evaluate unfiltered data
{wider_dt_all_resume <- dt_all_resume %>% 
  mutate(RRA = round(RRA,digits = 4)) %>% 
  ungroup() %>% 
  unite(Site, Level, Year, Filter, col= "Site_Level_Year_Filter") %>% 
  pivot_wider(id_cols = c("Class","Curated ID", "Final ID (BLASTn)", "Total reads", "Total abundance"),
              names_from = Site_Level_Year_Filter,
              values_from =  c("RRA","ASVs","OTUs","Reads"),
              names_glue = "{Site_Level_Year_Filter}_{.value}") %>%
  select(sort(colnames(.))) %>% 
  relocate("Class","Curated ID","Final ID (BLASTn)","Total reads", "Total abundance",
           #2020 MCE
           "Ponte_Full_2020_MCE_Reads",
           "Ponte_Full_2020_MCE_ASVs",
           "Ponte_Full_2020_MCE_OTUs",
           "Ponte_Full_2020_MCE_RRA",
           "Fundação_Full_2020_MCE_Reads",
           "Fundação_Full_2020_MCE_ASVs",
           "Fundação_Full_2020_MCE_OTUs",
           "Fundação_Full_2020_MCE_RRA",
           #2021 MCE
           "Prainha_Low_2021_MCE_Reads",
           "Prainha_Low_2021_MCE_ASVs",
           "Prainha_Low_2021_MCE_OTUs",
           "Prainha_Low_2021_MCE_RRA",
           "Barragem_Low_2021_MCE_Reads",
           "Barragem_Low_2021_MCE_ASVs",
           "Barragem_Low_2021_MCE_OTUs",
           "Barragem_Low_2021_MCE_RRA",
           "Ponte_Low_2021_MCE_Reads",
           "Ponte_Low_2021_MCE_ASVs",
           "Ponte_Low_2021_MCE_OTUs",
           "Ponte_Low_2021_MCE_RRA",
           "Fundação_Low_2021_MCE_Reads",
           "Fundação_Low_2021_MCE_ASVs",
           "Fundação_Low_2021_MCE_OTUs",
           "Fundação_Low_2021_MCE_RRA",
           #2021 Sterivex
           "Prainha_Low_2021_Sterivex_Reads",
           "Prainha_Low_2021_Sterivex_ASVs",
           "Prainha_Low_2021_Sterivex_OTUs",
           "Prainha_Low_2021_Sterivex_RRA",
           "Barragem_Low_2021_Sterivex_Reads",
           "Barragem_Low_2021_Sterivex_ASVs",
           "Barragem_Low_2021_Sterivex_OTUs",
           "Barragem_Low_2021_Sterivex_RRA",
           "Ponte_Low_2021_Sterivex_Reads",
           "Ponte_Low_2021_Sterivex_ASVs",
           "Ponte_Low_2021_Sterivex_OTUs",
           "Ponte_Low_2021_Sterivex_RRA",
           "Fundação_Low_2021_Sterivex_Reads",
           "Fundação_Low_2021_Sterivex_ASVs",
           "Fundação_Low_2021_Sterivex_OTUs",
           "Fundação_Low_2021_Sterivex_RRA",
           #2022 MCE
           "Prainha_Full_2022_MCE_Reads",
           "Prainha_Full_2022_MCE_ASVs",
           "Prainha_Full_2022_MCE_OTUs",
           "Prainha_Full_2022_MCE_RRA",
           "Barragem_Full_2022_MCE_Reads",
           "Barragem_Full_2022_MCE_ASVs",
           "Barragem_Full_2022_MCE_OTUs",
           "Barragem_Full_2022_MCE_RRA",
           "Ponte_Full_2022_MCE_Reads",
           "Ponte_Full_2022_MCE_ASVs",
           "Ponte_Full_2022_MCE_OTUs",
           "Ponte_Full_2022_MCE_RRA",
           "Fundação_Full_2022_MCE_Reads",
           "Fundação_Full_2022_MCE_ASVs",
           "Fundação_Full_2022_MCE_OTUs",
           "Fundação_Full_2022_MCE_RRA"
  )}

# View(wider_dt_all_resume)

write.csv(wider_dt_all_resume, paste0(tbl_path, "/wider_dt_all_resume_2024.csv"))

## Suplementar table B ----

# Creating the table with data to be used in analysis
# Filtered data, only Actinopteri and reads >=100

filt_tbl <- curated_full_tbl %>% 
  filter(`Class (BLASTn)` %in% "Actinopteri") %>% 
  group_by(`Curated ID`, Site, Level, Filter, Year) %>% 
  mutate(Reads = sum(Abundance)) %>% 
  filter(Reads >= 100) %>% 
  ungroup()

# View(filt_tbl)

# Filtered and grouped data by Samples
grouped_sampled_filt_tbl <- curated_full_tbl %>% 
  filter(`Class (BLASTn)` %in% "Actinopteri") %>% 
  group_by(`Curated ID`, Sample) %>% 
  mutate(Reads = sum(Abundance)) %>% 
  filter(Reads >= 100) %>% 
  ungroup() %>% 
  # creating total abundance by sample
  group_by(Sample) %>%
  mutate("Abd total" = sum(Abundance)) %>%
  ungroup() %>% 
  # calculating RRA by Sample
  group_by(`Curated ID`, Sample) %>%
  mutate("RRA na amostra" = (Abundance/`Abd total`*100)) %>% 
  summarize("Site" = first(Site),
          "Year" = first(Year),
          "Expedition" = first(Expedition),
          "Level" = first(Level),
          "Filter" = first(Filter),
          "Order" = first(`Order (BLASTn)`),
          "Family" = first(`Family (BLASTn)`),
          "Genus" = first(`Genus (BLASTn)`),
          "Reads" = sum(Abundance),
          "Sample abundance" = first(`Abd total`),
          "RRA" = sum(`RRA na amostra`),
          "ASVs" = length(unique(`ASV header`)),
          "OTUs" = length(unique(OTU)),
          ) 

# View(grouped_sampled_filt_tbl)

# Is RRA correct?
grouped_sampled_filt_tbl %>%
  group_by(Sample) %>%
  summarise(total_RRA = sum(RRA, na.rm = TRUE)) %>%
  ungroup()

# Filtered and grouped data by Year, Site and Filter
grouped_filt_tbl <- filt_tbl %>% 
  # creating total abundance by period
  group_by(Level, Site, Filter, Year) %>%
  mutate("Abd total" = sum(Abundance)) %>%
  ungroup() %>%
  # creating total abundance by Curated ID
  group_by(`Curated ID`) %>% 
  mutate("Abundancia total" = sum(Abundance)) %>% 
  ungroup() %>% 
  # calculating RRA by period
  group_by(`Curated ID`, Site, Level, Filter, Year) %>%
  mutate("RRA no periodo" = (Abundance/`Abd total`)*100) %>% 
  reframe("Curated ID" = `Curated ID`,
          "Order" = `Order (BLASTn)`,
          "Family" = `Family (BLASTn)`,
          "Genus" = `Genus (BLASTn)`,
          "Reads" = sum(Abundance),
          "Period abundance" = `Abd total`,
          "Total abundance" = `Abundancia total`, 
          "RRA" = sum(`RRA no periodo`),
          "ASVs" = length(unique(`ASV header`)),
          "OTUs" = length(unique(OTU))) %>% 
  unique()

# View(grouped_filt_tbl_tbl)

# Is RRA correct?
grouped_filt_tbl %>%
  group_by(Site, Level, Filter, Year) %>%
  summarise(total_RRA = sum(RRA, na.rm = TRUE)) %>%
  ungroup()

# Tabela wider com os dados filtrados
{
  wider_filt_tbl <- grouped_filt_tbl %>% 
  mutate(RRA = round(RRA,digits = 4)) %>% 
  ungroup() %>% 
  unite(Site, Level, Year, Filter, col= "Site_Level_Year_Filter") %>% 
  pivot_wider(id_cols = c("Curated ID", "Total abundance"),
              names_from = Site_Level_Year_Filter,
              values_from =  c("RRA","ASVs","OTUs","Reads"),
              names_glue = "{Site_Level_Year_Filter}_{.value}") %>%
  select(sort(colnames(.))) %>% 
  relocate("Curated ID",
           "Total abundance",
           #2020 MCE
           "Ponte_Full_2020_MCE_Reads", 
           "Ponte_Full_2020_MCE_ASVs", 
           "Ponte_Full_2020_MCE_OTUs",
           "Ponte_Full_2020_MCE_RRA",
           "Fundação_Full_2020_MCE_Reads",
           "Fundação_Full_2020_MCE_ASVs",
           "Fundação_Full_2020_MCE_OTUs",
           "Fundação_Full_2020_MCE_RRA",
           #2021 MCE
           "Prainha_Low_2021_MCE_Reads",
           "Prainha_Low_2021_MCE_ASVs",
           "Prainha_Low_2021_MCE_OTUs",
           "Prainha_Low_2021_MCE_RRA",
           "Barragem_Low_2021_MCE_Reads",
           "Barragem_Low_2021_MCE_ASVs",
           "Barragem_Low_2021_MCE_OTUs",
           "Barragem_Low_2021_MCE_RRA",
           "Ponte_Low_2021_MCE_Reads",
           "Ponte_Low_2021_MCE_ASVs",
           "Ponte_Low_2021_MCE_OTUs",
           "Ponte_Low_2021_MCE_RRA",
           "Fundação_Low_2021_MCE_Reads",
           "Fundação_Low_2021_MCE_ASVs",
           "Fundação_Low_2021_MCE_OTUs",
           "Fundação_Low_2021_MCE_RRA",
           #2021 Sterivex
           "Prainha_Low_2021_Sterivex_Reads",
           "Prainha_Low_2021_Sterivex_ASVs",
           "Prainha_Low_2021_Sterivex_OTUs",
           "Prainha_Low_2021_Sterivex_RRA",
           "Barragem_Low_2021_Sterivex_Reads",
           "Barragem_Low_2021_Sterivex_ASVs",
           "Barragem_Low_2021_Sterivex_OTUs",
           "Barragem_Low_2021_Sterivex_RRA",
           "Ponte_Low_2021_Sterivex_Reads",
           "Ponte_Low_2021_Sterivex_ASVs",
           "Ponte_Low_2021_Sterivex_OTUs",
           "Ponte_Low_2021_Sterivex_RRA",
           "Fundação_Low_2021_Sterivex_Reads",
           "Fundação_Low_2021_Sterivex_ASVs",
           "Fundação_Low_2021_Sterivex_OTUs",
           "Fundação_Low_2021_Sterivex_RRA",
           #2022 MCE
           "Prainha_Full_2022_MCE_Reads",
           "Prainha_Full_2022_MCE_ASVs",
           "Prainha_Full_2022_MCE_OTUs",
           "Prainha_Full_2022_MCE_RRA",
           "Barragem_Full_2022_MCE_Reads",
           "Barragem_Full_2022_MCE_ASVs",
           "Barragem_Full_2022_MCE_OTUs",
           "Barragem_Full_2022_MCE_RRA"
           # "Ponte_Full_2022_MCE_Reads", # Os dois prox pontos nao sobreviveram 
           # "Ponte_Full_2022_MCE_ASVs", # aos filtros por que nao tem ASVs de peixes 
           # "Ponte_Full_2022_MCE_OTUs",
           # "Ponte_Full_2022_MCE_RRA",
           # "Fundação_Full_2022_MCE_Reads",
           # "Fundação_Full_2022_MCE_ASVs",
           # "Fundação_Full_2022_MCE_OTUs",
           # "Fundação_Full_2022_MCE_RRA"
  )
  }

# View(wider_filt_tbl)

write.csv(wider_filt_tbl, paste0(tbl_path, "/wider_dt_filt_2024.csv"))
```


## Basic stats
```{r, eval=FALSE,echo=FALSE}
options(scipen = 999)

# Reads, ASVs and Species by Curated ID and Order ----

## Raw stats

curated_full_tbl %>% 
  group_by(`Class (BLASTn)`) %>% 
  reframe(
    "Total reads" = sum(Abundance), 
    "ASVs" = n_distinct(`ASV (Sequence)`),
    "OTUs" = n_distinct(OTU),
    "Orders" = length(unique(`Order (BLASTn)`)),
    "Families" = length(unique(`Family (BLASTn)`)),
    "Genus" = length(unique(`Genus (BLASTn)`)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & !grepl("sp\\.", `Curated ID`)])),
    "Nspecies" = length(unique(`Curated ID`)) - Species
  ) %>% 
  unique() %>% 
  View()

## Stats Actinopteri
grouped_sampled_filt_tbl %>% 
  View()

grouped_sampled_filt_tbl %>% 
  group_by(`Curated ID`) %>% 
  mutate("Total reads" = sum(Reads)) %>%
  ungroup() %>% 
  reframe(
    # "Total reads" = sum(`Total reads`),
    "Total reads" = `Total reads`,
    "ASVs" = n_distinct(`ASV header`),
    "OTUs" = n_distinct(`OTU number`),
    "Orders" = length(unique(Order)),
    "Families" = length(unique(Family)),
    "Genus" = length(unique(Genus)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & !grepl("sp\\.", `Curated ID`)])),
    "Nspecies" = length(unique(`Curated ID`)) - Species
  )  %>% 
  unique() %>% 
  View()

## Grouped by Curated ID
grouped_sampled_filt_tbl %>% 
  mutate("Total reads" = sum(Reads)) %>% 
  group_by(`Curated ID`) %>% 
  reframe("Total species reads" = sum(Reads),
            "Relative abundance for species" = `Total species reads` / `Total reads` * 100,
          "Order" = Order) %>% 
  relocate(Order, `Curated ID`) %>% 
  unique() %>% 
  View()

## Grouped by Order
grouped_sampled_filt_tbl %>% 
  mutate("Total reads" = sum(Reads)) %>% 
  group_by(Order) %>% 
  reframe("Total order reads" = sum(Reads),
          "Relative abundance for order" = `Total order reads` / `Total reads` * 100,
          "Total ASVs" = sum(ASVs),
          "Total Species" = n_distinct(`Curated ID`)) %>% 
  unique() %>% 
  View()

# Reads, ASVs and Species by Filter ----

## Grouped by Filter
grouped_sampled_filt_tbl %>% 
  filter(Expedition %in% "Novembro 2021") %>%
  mutate("Total reads" = sum(Reads)) %>% 
  group_by(Filter, 
           # `Curated ID`
           ) %>% 
  reframe("Total filter reads" = sum(Reads),
          "Relative abundance for filter" = `Total filter reads` / `Total reads` * 100,
          "Total ASVs" = sum(ASVs),
          "Total orders" =  n_distinct(Order),
          "Total families" =  n_distinct(Family),
          "Total Species" = n_distinct(`Curated ID`),
          "Total Samples" = n_distinct(Sample)) %>% 
  unique() %>% 
  View()


```


## Exploratory tables

```{r, eval=FALSE,echo=FALSE}

## Exploratory tables ----

# Reads through data filtering 
raw_results_reads <- raw_results_tbl %>% group_by(Sample) %>% 
  summarise(`Raw reads` = sum(Abundance))

pre_filt_reads <- pre_filt_tbl %>% group_by(Sample) 

filt_reads <- filt_tbl %>% group_by(Sample) %>% 
  summarise(`Filt reads` = sum(Abundance))

evol_reads <- raw_results_reads %>%
  left_join(pre_filt_reads, by = "Sample") %>%
  left_join(filt_reads, by = "Sample") %>% 
  group_by(`Raw reads`, `Filt reads`) %>% 
  select(Sample, `Raw reads`, `Filt reads`) %>% 
  unique()

View(evol_reads)

# ASVs that weren't identified at species Level (genus + epithet)
spp_level <- grouped_filt_tbl %>%
  reframe("Species" = unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & 
                                            !grepl("sp\\.", `Curated ID`)]))
nspp_level <- grouped_filt_tbl %>% 
  reframe("Nspecies" = setdiff(unique(`Curated ID`), spp_level))
    
View(nspp_level) 

# Informations grouped by taxa detected 
id_level <- grouped_filt_tbl_tbl %>%
  group_by(`Curated ID`) %>%
    reframe("Order" = Order, "Family" = Family, "Genus" = Genus,
      "Curated identification" = `Curated ID`,
      "Reads" = sum(Reads), 
      "ASVs" = sum(ASVs),
      "OTUs" = sum(OTUs),
    ) %>% 
    unique() %>% 
    select("Order", "Family", "Genus",
           # "Maior hit do BLASTn",
           "Curated identification",
           "Reads", "ASVs","OTUs") 

View(id_level)

# Informations grouped  by Class
class_Level <- grouped_filt_tbl %>%
  reframe(
    "Total reads" = sum(Reads), 
    "ASVs" = sum(ASVs),
    "OTUs" = sum(OTUs),
    "Orders" = length(unique(Order)),
    "Families" = length(unique(Family)),
    "Genera" = length(unique(Genus)),
    "Identifications" = length(unique(`Curated ID`)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & 
                                               !grepl("sp\\.", `Curated ID`)])), # IDs a Level de spp
    "Nspecies" = length(unique(`Curated ID`)) - Species) # IDs em outros niveis taxonomicos
  
View(class_Level)
  
# Informations grouped by Order
order_level <- grouped_filt_tbl %>%
  group_by(Order) %>%
  reframe(
    "Order" = Order,
    "Total reads" = sum(Reads), # dt_all_resume && # dt_filt
    "ASVs" = sum(ASVs),
    "OTUs" = sum(OTUs),
    "Families" = length(unique(Family)),
    "Genera" = length(unique(Genus)),
    "Identifications" = length(unique(`Curated ID`)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & !grepl("sp\\.", `Curated ID`)])), # a nivel de spp
    "Nspecies" = length(unique(`Curated ID`)) - Species # em outros niveis alem de spp
    ) %>% 
  unique()
  
# View(order_level)

# Informations grouped by Family
family_level <- grouped_filt_tbl %>%
  group_by(Family) %>%
  reframe(
    "Total reads" = sum(Reads), 
    "ASVs" = sum(ASVs),
    "OTUs" = sum(OTUs),
    "Genera" = length(unique(Genus)),
    "Identifications" = length(unique(`Curated ID`)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & !grepl("sp\\.", `Curated ID`)])), # a nivel de spp
    "Nspecies" = length(unique(`Curated ID`)) - Species # em outros niveis alem de spp
    ) %>% 
  unique()
  
# View(family_level)

# Informations grouped by Sample
grouped_by_Sample <- filt_tbl %>%
 # filter(Expedition %in% "Novembro 2021") %>%
 select(Sample, Filter, Reads, Site) %>%
 group_by(Filter,
          Sample
          ) %>%
 mutate("Total reads" = sum(Reads)) %>%
 reframe(Filter,
         # Sample,
         Site,
         `Total reads`) %>%
 unique()

# View(grouped_by_Sample)

# Especies encontradas em apenas uma amostra
Counts <- grouped_filt_tbl %>% 
  group_by(`Curated ID`) %>%
  # group_by(`Curated ID`, Filter) %>%
  reframe(count = n_distinct(Sample),
            Filter, Sample) %>% 
  unique()

unique_counts <- 
  Counts %>% filter(count == 1) %>% 
  reframe(`Curated ID`,
          Sample)

View(unique_counts)
```

## Exploratory plots

```{r, eval=FALSE,echo=FALSE}
## Plots ----

# Proportions of identified orders 
{fish_ord_all <- grouped_filt_tbl %>% 
  unite(Site, Year, Filter, sep = "_", remove = FALSE, col = "un_amostral") %>% 
  select(c( Order,`Curated ID`, Site, Year, Filter, un_amostral)) %>% 
  unique() %>%
  group_by(Site, Year, Filter) %>% 
  mutate("alpha_d" = length(`Curated ID`)) %>%
  ungroup() %>% 
  group_by(Order,Site, Year, Filter) %>% 
  mutate("alpha_d_order" = length(`Curated ID`)) %>% 
  ungroup() %>% 
  mutate("alpha_d_order_%" = round(alpha_d_order/alpha_d*100, digits = 2)) %>% 
  select(-c(`Curated ID`)) %>% 
  unique() %>% 
  group_by(un_amostral) %>% 
  mutate(Year_Level = case_when(Year == "2020" ~ "2020 - Full",
                                Year == "2021" ~ "2021 - Low",
                                Year == "2022" ~ "2022 - Full")) %>%
  filter(Filter %in% "MCE") %>% 
  ggplot(aes(x = Site,
             y = alpha_d_order,
             fill = Order)) +
  geom_col(position = "fill") +
  geom_text(aes(label = sprintf("%0.1f%%",`alpha_d_order_%`)), # adicionar % dentro das barras
            position = position_fill(vjust = 0.5),
            color = "black") +
  facet_grid(rows = vars(Year_Level),
             space = "free",
             scales = "free",
             drop = TRUE) +
  coord_flip() + #virando 90o o grafico
  guides(fill = guide_legend("Orders")) + ## alterar o titulo da legenda 
  theme(panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2))
        ) +
  theme(plot.title = element_text(size = 20, face = "bold"),
        axis.text.x = element_text(size = 20), ## definindo o tamanho de cada texto
        axis.text.y = element_text(size = 20),
        axis.title.x = element_text(size = 20, face = "bold"),
        axis.title.y = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 19),
        legend.title = element_text(size = 19),
        strip.text = element_text(size = 15, face = "bold"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "lightgray", color = NA)) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("#5d9cecff",
                               "#4fc1e9ff",
                               "#ac92ecff",
                               "#a0d468ff",
                               "#ffce54ff",
                               "#fc6e51ff",
                               "#ed5565ff",
                               "#48cfadff",                                  
                               "#ec87c0ff")) +
  labs(x = "Sample sites", y = "Proportion") 

fish_ord_all

ggsave(plot = fish_ord_all, 
       filename = paste0(figs_path, "/alpha_ord", "-", Sys.Date(), ".pdf", sep = ""),
       device = "pdf", units = "cm", height = 20, width = 40, dpi = 600)}

# Proportions of identified orders HoH suggestions
{fish_ord_all <- grouped_filt_tbl %>% 
  unite(Site, Year, Filter, sep = "_", remove = FALSE, col = "un_amostral") %>% 
  select(c( Order,`Curated ID`, Site, Year, Filter, un_amostral)) %>% 
  unique() %>%
  group_by(Site, Year, Filter) %>% 
  mutate("alpha_d" = length(`Curated ID`)) %>%
  ungroup() %>% 
  group_by(Order,Site, Year, Filter) %>% 
  mutate("alpha_d_order" = length(`Curated ID`)) %>% 
  ungroup() %>% 
  mutate("alpha_d_order_%" = round(alpha_d_order/alpha_d*100, digits = 2)) %>% 
  select(-c(`Curated ID`)) %>% 
  unique() %>% 
  group_by(un_amostral) %>% 
  mutate(Year_Level = case_when(Year == "2020" ~ "2020 - Full",
                                Year == "2021" ~ "2021 - Low",
                                Year == "2022" ~ "2022 - Full")) %>%
  mutate(Year_Level = factor(Year_Level, levels = c("2022 - Full",
                                                    "2021 - Low",
                                                    "2020 - Full"))) %>%
  filter(Filter %in% "MCE") %>% 
  ggplot(aes(
             # x = Site,
             x = Year_Level,
             y = alpha_d_order,
             fill = Order)) +
  geom_col(position = "fill") +
  geom_text(aes(label = sprintf("%0.1f%%",`alpha_d_order_%`)), # adicionar % dentro das barras
            position = position_fill(vjust = 0.5),
            color = "black") +
  # facet_grid(rows = vars(Year_Level),
  facet_grid(rows = vars(Site),
  space = "free", 
  scales = "free",
  drop = TRUE) +
  coord_flip() + #virando 90o o grafico
  guides(fill = guide_legend("Orders")) + ## alterar o titulo da legenda 
  theme(panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2))
        ) +
  theme(plot.title = element_text(size = 20, face = "bold"),
        axis.text.x = element_text(size = 20), ## definindo o tamanho de cada texto
        axis.text.y = element_text(size = 20),
        axis.title.x = element_text(size = 20, face = "bold"),
        axis.title.y = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 19),
        legend.title = element_text(size = 19),
        strip.text = element_text(size = 15, face = "bold"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "lightgray", color = NA)) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("#5d9cecff",
                               "#4fc1e9ff",
                               "#ac92ecff",
                               "#a0d468ff",
                               "#ffce54ff",
                               "#fc6e51ff",
                               "#ed5565ff",
                               "#48cfadff",                                  
                               "#ec87c0ff")) +
  labs(x = "Year and water level", y = "Proportion") 

fish_ord_all

ggsave(plot = fish_ord_all, 
       filename = paste0(figs_path, "/alpha_ord_HoH", "-", Sys.Date(), ".pdf", sep = ""),
       device = "pdf", units = "cm", height = 20, width = 40, dpi = 600)
}

# Proportions of identified families
{fish_fam_all <- grouped_filt_tbl %>%
  unite(Site, Year, Filter,sep = "_", remove = FALSE, col = "un_amostral") %>% 
  select(c(Order, Family,`Curated ID`, Site, Year, Filter, un_amostral)) %>% 
  unique() %>%
  group_by(Site, Year, Filter) %>% 
  mutate("alpha_d" = length(`Curated ID`)) %>%
  ungroup() %>% 
  group_by(Family,Site, Year, Filter) %>% 
  mutate("alpha_d_order" = length(`Curated ID`)) %>% 
  ungroup() %>% 
  mutate("alpha_d_order_%" = round(alpha_d_order/alpha_d*100, digits = 2)) %>% 
  select(-c(`Curated ID`)) %>% 
  unique() %>% 
  group_by(un_amostral) %>% 
  mutate(Year_Level = case_when(Year == "2020" ~ "2020 - Full",
                                Year == "2021" ~ "2021 - Low",
                                Year == "2022" ~ "2022 - Full")) %>%
  filter(Filter %in% "MCE") %>% 
  ggplot(aes(x = Site,
             y = alpha_d_order,
             fill = Family)) +
  geom_col(position = "fill") +
  geom_text(aes(label = sprintf("%0.1f%%",`alpha_d_order_%`)), #adicionando % dentro das barras
            position = position_fill(vjust = 0.5),
            color = "black") +
  facet_grid(rows = vars(Year_Level),
  space = "free", 
  scales = "free",
  drop = TRUE) +
  coord_flip() + #virando 90o o grafico
  guides(fill = guide_legend("Family")) + ## alterar o titulo da legenda 
  theme(panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2))
        ) +
  theme(plot.title = element_text(size = 20, face = "bold"),
        axis.text.x = element_text(size = 20), ## definindo o tamanho de cada texto
        axis.text.y = element_text(size = 20),
        axis.title.x = element_text(size = 20, face = "bold"),
        axis.title.y = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 19),
        legend.title = element_text(size = 19),
        strip.text = element_text(size = 15, face = "bold"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "lightgray", color = NA)) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("#5d9cecff",
                               "#4fc1e9ff",
                               "#48cfadff",
                               "#a0d468ff",
                               "#ffce54ff",
                               "#fc6e51ff",
                               "#ed5565ff",
                               "#ac92ecff",
                               "#ec87c0ff"
                               )) +
  labs(x = "Sample sites", y = "Proportion")

fish_fam_all

ggsave(plot = fish_fam_all, 
         filename = paste0(figs_path, "/alpha_fam_filt", "-", Sys.Date(), ".pdf", sep = ""),
         device = "pdf", units = "cm", height = 21, width = 40, dpi = 600)
}

# Proportions of identified families HoH suggestions 
{fish_fam_all <- grouped_filt_tbl %>%
  unite(Site, Year, Filter,sep = "_", remove = FALSE, col = "un_amostral") %>% 
  select(c(Order, Family,`Curated ID`, Site, Year, Filter, un_amostral)) %>% 
  unique() %>%
  group_by(Site, Year, Filter) %>% 
  mutate("alpha_d" = length(`Curated ID`)) %>%
  ungroup() %>% 
  group_by(Family,Site, Year, Filter) %>% 
  mutate("alpha_d_order" = length(`Curated ID`)) %>% 
  ungroup() %>% 
  mutate("alpha_d_order_%" = round(alpha_d_order/alpha_d*100, digits = 2)) %>% 
  select(-c(`Curated ID`)) %>% 
  unique() %>% 
  group_by(un_amostral) %>% 
  mutate(Year_Level = case_when(Year == "2020" ~ "2020 - Full",
                                Year == "2021" ~ "2021 - Low",
                                Year == "2022" ~ "2022 - Full")) %>%
  mutate(Year_Level = factor(Year_Level, levels = c("2022 - Full",
                                                    "2021 - Low",
                                                    "2020 - Full"))) %>%
  filter(Filter %in% "MCE") %>% 
  ggplot(aes(x = Year_Level,
             y = alpha_d_order,
             fill = Family)) +
  geom_col(position = "fill") +
  geom_text(aes(label = sprintf("%0.1f%%",`alpha_d_order_%`)), #adicionando % dentro das barras
            position = position_fill(vjust = 0.5),
            color = "black") +
  # facet_grid(rows = vars(Year_Level),
  facet_grid(rows = vars(Site),
  space = "free", 
  scales = "free",
  drop = TRUE) +
  coord_flip() + #virando 90o o grafico
  guides(fill = guide_legend("Family")) + ## alterar o titulo da legenda 
  theme(panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2))
        ) +
  theme(plot.title = element_text(size = 20, face = "bold"),
        axis.text.x = element_text(size = 20), ## definindo o tamanho de cada texto
        axis.text.y = element_text(size = 20),
        axis.title.x = element_text(size = 20, face = "bold"),
        axis.title.y = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 19),
        legend.title = element_text(size = 19),
        strip.text = element_text(size = 15, face = "bold"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "lightgray", color = NA)) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("#5d9cecff",
                               "#4fc1e9ff",
                               "#48cfadff",
                               "#a0d468ff",
                               "#ffce54ff",
                               "#fc6e51ff",
                               "#ed5565ff",
                               "#ac92ecff",
                               "#ec87c0ff"
                               )) +
  labs(x = "Year and water level", y = "Proportion")

fish_fam_all

ggsave(plot = fish_fam_all, 
         filename = paste0(figs_path, "/alpha_fam_filt_HoH", "-", Sys.Date(), ".pdf", sep = ""),
         device = "pdf", units = "cm", height = 21, width = 40, dpi = 600)
}

# Proportions of species identifications
{id_prop_MCE_STX <- filt_tbl %>%
  unite(Site, Year, Filter, sep = "_", remove = FALSE, col = "un_amostral") %>% 
  filter(Expedition %in% "Novembro 2021") %>% 
  select(c(`Curated ID`, Site, Year, Filter, un_amostral)) %>% 
  unique() %>%
  group_by(Site, Year, Filter) %>% 
  mutate("alpha_d" = length(`Curated ID`)) %>%
  ungroup() %>% 
  group_by(`Curated ID`, Site, Year, Filter) %>% 
  mutate("alpha_d_order" = length(`Curated ID`)) %>% 
  ungroup() %>% 
  mutate("alpha_d_order_%" = round(alpha_d_order / alpha_d * 100, digits = 2)) %>% 
  unique() %>% 
  # group_by(un_amostral) %>% 
  # mutate(Year_Level = case_when(Year == "2020" ~ "2020 - Full",
  #                               Year == "2021" ~ "2021 - Low",
  #                               Year == "2022" ~ "2022 - Full")) %>%
  # filter(Filter %in% "MCE") %>% 
  # filter(Expedition %in% "Novembro 2021") %>%
  ggplot(aes(x = Site,
             y = alpha_d_order,
             fill = `Curated ID`)) +
  geom_col(position = "fill") +
  geom_text(aes(label = sprintf("%0.1f%%",`alpha_d_order_%`)), #adicionando % dentro das barras
            position = position_fill(vjust = 0.5),
            color = "black") +
  facet_grid(rows = vars(Filter),
  space = "free", 
  scales = "free",
  drop = TRUE) +
  coord_flip() + #virando 90o o grafico
  guides(fill = guide_legend("Curated ID")) + ## alterar o titulo da legenda 
  theme(panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2))
        ) +
  theme(plot.title = element_text(size = 20, face = "bold"),
        axis.text.x = element_text(size = 20), ## definindo o tamanho de cada texto
        axis.text.y = element_text(size = 20),
        axis.title.x = element_text(size = 20, face = "bold"),
        axis.title.y = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 19),
        legend.title = element_text(size = 19),
        strip.text = element_text(size = 15, face = "bold"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "lightgray", color = NA)) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("#1f77b4",  # Azul
                               "#ff7f0e",  # Laranja
                               "#2ca02c",  # Verde
                               "#d62728",  # Vermelho
                               "#9467bd",  # Roxo
                               "#8c564b",  # Marrom
                               "#e377c2",  # Rosa
                               "#7f7f7f",  # Cinza
                               "#bcbd22",  # Verde-amarelado
                               "#17becf",  # Ciano
                               "#fdae61",  # Laranja-claro
                               "#ff9896",  # Rosa-pálido
                               "#c49c94",  # Marrom-claro
                               "#aec7e8",  # Azul-claro
                               "#ffbb78",  # Laranja-suave
                               "#98df8a",  # Verde-pálido
                               "#c5b0d5",  # Lilás
                               "#ff9896",  # Rosa-suave
                               "#dbdb8d",  # Amarelo-pálido
                               "#9edae5",  # Azul-petróleo claro
                               "#ad494a",  # Vinho
                               "#d53e4f",  # Rosa-forte
                               "#393b79",  # Azul-marinho
                               "#637939",  # Verde-musgo
                               "#8c6d31",  # Dourado-escuro
                               "#66c2a5",   # Verde-turquesa
                               "#5254a3",  # Azul-índigo
                               "#6b6ecf",  # Azul-púrpura
                               "#bd9e39"   # Mostarda-claro
                               )) +
  labs(x = "Sample sites", y = "Proportion")

id_prop_MCE_STX

ggsave(plot = id_prop,
         filename = paste0(figs_path, "/alpha_ids_MCE-STX_prop", "-", Sys.Date(), ".pdf", sep = ""),
         device = "pdf", units = "cm", height = 22.5, width = 50, dpi = 600)
}

# Proportions of species identifications
{id_prop_MCE_STX <- filt_tbl %>%
  unite(Site, Year, Filter, sep = "_", remove = FALSE, col = "un_amostral") %>% 
  filter(Expedition %in% "Novembro 2021") %>%
  select(c(`Curated ID`, Site, Year, Filter, un_amostral)) %>% 
  unique() %>%
  group_by(Site, Year, Filter) %>% 
  mutate("alpha_d" = length(`Curated ID`)) %>%
  ungroup() %>% 
  group_by(`Curated ID`, Site, Year, Filter) %>% 
  mutate("alpha_d_order" = length(`Curated ID`)) %>% 
  ungroup() %>% 
  mutate("alpha_d_order_%" = round(alpha_d_order / alpha_d * 100, digits = 2)) %>% 
  unique() %>% 
  group_by(un_amostral) %>%
  mutate(Year_Level = case_when(Year == "2020" ~ "2020 - Full",
                                Year == "2021" ~ "2021 - Low",
                                Year == "2022" ~ "2022 - Full")) %>%
  mutate(Year_Level = factor(Year_Level, levels = c("2022 - Full",
                                                    "2021 - Low",
                                                    "2020 - Full"))) %>%  
  # filter(Filter %in% "MCE") %>%
  # filter(Expedition %in% "Novembro 2021") %>%
  ggplot(aes(x = Filter,
             # x = Site,
             # x = Year_Level,
             y = alpha_d_order,
             fill = `Curated ID`)) +
  geom_col(position = "fill") +
  geom_text(aes(label = sprintf("%0.1f%%",`alpha_d_order_%`)), #adicionando % dentro das barras
            position = position_fill(vjust = 0.5),
            color = "black") +
  facet_grid(rows = vars(Site),
  space = "free", 
  scales = "free",
  drop = TRUE) +
  coord_flip() + #virando 90o o grafico
  guides(fill = guide_legend("Curated ID")) + ## alterar o titulo da legenda 
  theme(panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2))
        ) +
  theme(plot.title = element_text(size = 20, face = "bold"),
        axis.text.x = element_text(size = 20), ## definindo o tamanho de cada texto
        axis.text.y = element_text(size = 20),
        axis.title.x = element_text(size = 20, face = "bold"),
        axis.title.y = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 19),
        legend.title = element_text(size = 19),
        strip.text = element_text(size = 15, face = "bold"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "lightgray", color = NA)) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("#1f77b4",  # Azul
                               "#ff7f0e",  # Laranja
                               "#2ca02c",  # Verde
                               "#d62728",  # Vermelho
                               "#9467bd",  # Roxo
                               "#8c564b",  # Marrom
                               "#e377c2",  # Rosa
                               "#7f7f7f",  # Cinza
                               "#bcbd22",  # Verde-amarelado
                               "#17becf",  # Ciano
                               "#fdae61",  # Laranja-claro
                               "#ff9896",  # Rosa-pálido
                               "#c49c94",  # Marrom-claro
                               "#aec7e8",  # Azul-claro
                               "#ffbb78",  # Laranja-suave
                               "#98df8a",  # Verde-pálido
                               "#c5b0d5",  # Lilás
                               "#ff9896",  # Rosa-suave
                               "#dbdb8d",  # Amarelo-pálido
                               "#9edae5",  # Azul-petróleo claro
                               "#ad494a",  # Vinho
                               "#d53e4f",  # Rosa-forte
                               "#393b79",  # Azul-marinho
                               "#637939",  # Verde-musgo
                               "#8c6d31",  # Dourado-escuro
                               "#66c2a5",   # Verde-turquesa
                               "#5254a3",  # Azul-índigo
                               "#6b6ecf",  # Azul-púrpura
                               "#bd9e39"   # Mostarda-claro
                               )) +
  labs(x = "Filter", y = "Proportion")

id_prop_MCE_STX

ggsave(plot = id_prop_MCE_STX,
         filename = paste0(figs_path, "/alpha_ids_MCE-STX_prop_HoH", "-", Sys.Date(), ".pdf", sep = ""),
         device = "pdf", units = "cm", height = 22.5, width = 50, dpi = 600)
}

# Tile plots 
{
  # Defining levels 
  {
    # Defining order levels
    order_levels <- grouped_filt_tbl$Order %>% 
      sort() %>% unique()
    
    # Defining spp levels
    spp_levels <- grouped_filt_tbl$`Curated ID` %>% 
      sort() %>% unique()
    
    # Defining sample levels
    grouped_filt_tbl$Sample %>% unique()
    
    samples <- c("L1_nov_dec_20_mi",
                 "L2_dez20",
                 "L2_nov20",
                 "L1_out21",
                 "L2_out21",
                 "L3_out21",
                 "L4_out21",
                 "L1_nov21",
                 "L2_nov21",
                 "L3_nov21",
                 "L4_nov21",
                 "STX_L1_nov21",
                 "STX_L2_nov21",
                 "STX_L3_nov21",
                 "STX_L4_nov21",
                 "L1_jan22",
                 "L2_jan22",
                 "L3_jan22",
                 "L4_jan22"
    )
    
    # Defining Expedition levels
    grouped_filt_tbl$Expedition %>% unique()
    
    Expeditions <- c("Novembro e Dezembro 2020",
                     "Novembro 2020",
                     "Dezembro 2020",
                     "Outubro 2021",
                     "Novembro 2021",
                     "Janeiro 2022")
    
  }
  
  # Tile plots 
  {
    # All Years (only MCE Filters) HoH suggestions
    tile_all <- grouped_filt_tbl %>%
      mutate(`Curated ID` = factor(`Curated ID`, levels = rev(spp_Levels))) %>%
      # mutate(Expedition = factor(Expedition, levels = Expeditions)) %>%
      # mutate(Sample = factor(Sample, levels = samples)) %>% 
      mutate(Site = factor(Site)) %>% 
      mutate(Order = factor(Order)) %>%
      mutate(Year_Level = case_when(
        Year == "2020" ~ "2020 - Full",
        Year == "2021" ~ "2021 - Empty",
        Year == "2022" ~ "2022 - Full"
      )) %>%
      # group_by(Expedition, `Curated ID`, Site, Order, Level) %>%   # Agrupa por expedição, Curated ID e Point
      filter(Filter %in% c("MCE")) %>%
      ggplot(aes(y = `Curated ID`, 
                 group = `Curated ID`,
                 # x = Site,
                 x = Year_Level,
                 fill = RRA)) +
      geom_tile() +
      # geom_text(aes(label= sprintf("%0.2f", round(RRA, digits = 2))), # exibindo os valores de RRA dentro dos tiles para facilitar a discussao (opcional)
      # stat = "identity",
      # colour = "black", size = 4) +
      facet_grid(cols = vars(Site),
                 # cols = vars(Year_Level),
                 rows = vars(Order),
                 space = "free", 
                 scales = "free",
                 drop = TRUE) +
      scale_fill_gradientn(name = "RRA (%)",
                           # colours = c("white", "yellow", "green", "darkgreen", "blue"),
                           colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF",
                                           "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
                           # colours = c("white", "#FDE725FF", "#6DCD59FF", "#35B779FF", "#3E4A89FF"),
                           # colours = c("#440154FF", "#482878FF", "#3E4A89FF", "#31688EFF",
                           #             "#26828EFF", "#1F9E89FF", "#35B779FF", "#6DCD59FF",
                           #             "#B4DE2CFF", "#FDE725FF"),
                           values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
                           breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00), # Quebras exatas
                           labels = function(x) scales::number(x, accuracy = 0.01, trim = FALSE), # Precisão exata
                           limits = c(0.01, 100),
                           na.value = "#7A0403FF",
                           trans = "log10"
      ) +
      scale_colour_gradientn(name = "RRA (%)",
                             # colours = rev(c("white", "yellow", "green", "darkgreen", "blue")),
                             colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF",
                                             "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
                             # colours = rev(c("white", "#FDE72C5FF", "#6DCD59FF", "#35B779FF", "#3E4A89FF")),
                             # colours = c("#440154FF", "#482878FF", "#3E4A89FF", "#31688EFF",
                             #             "#26828EFF", "#1F9E89FF", "#35B779FF", "#6DCD59FF",
                             #             "#B4DE2CFF", "#FDE725FF"),
                             values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
                           breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00), # Quebras exatas
                             labels = function(x) scales::number(x, accuracy = 0.01, trim = FALSE), # Precisão exata
                             limits = c(0.01, 100),
                             na.value = "#7A0403FF",
                             trans = "log10") +
      theme_minimal() +
      theme(panel.background = element_blank(), 
            panel.grid.major = element_line(color = "grey",
                                            linewidth = 0.2,
                                            linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2)),
        strip.background = element_rect(fill = "#e4e4e4"),
        strip.text = element_text(color = "black", size = rel(1.2))) +
      theme(plot.title = element_text(size = 20, face = "bold"),
            axis.text.x = element_text(size = 14, angle = 20, hjust = 1, vjust = 1),
            # axis.text.x = element_text(size = 14, vjust = 1),
            axis.text.y = element_text(size = 14, face = "italic"),
            axis.title.x = element_text(size = 14, face = "bold"),
            axis.title.y = element_text(size = 14, face = "bold"),
            strip.text = element_text(size = 14, face = "bold"),
            strip.text.y = element_text(angle=0),
            legend.text = element_text(size = 13),
            legend.title = element_text(size = 14, face = "bold", margin = margin(b = 10)),
            legend.position = "right",
            legend.key.height = unit(3, 'cm')
      ) +
      labs(fill ='RRA (%)',
           x = "Sample sites",
           y = "Species"
      )
    
    tile_all
    
    ggsave(plot = tile_all,
           filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
                            "tile_20-22_MCE_HoH", "-", Sys.Date(), ".pdf", sep = ""),
           device = "pdf", units = "cm", height = 20, width = 30, dpi = 600)
    
    # MCE versus Sterivex HoH suggestions
    tile_mce_vs_sterivex <- grouped_sampled_filt_tbl %>%
      mutate(`Curated ID` = factor(`Curated ID`, levels = rev(spp_levels))) %>%
      mutate(Site = factor(Site)) %>% 
      mutate(Order = factor(Order)) %>%
      mutate(Year_Level = case_when(
        Year == "2020" ~ "2020 - Full",
        Year == "2021" ~ "2021 - Low",
        Year == "2022" ~ "2022 - Full"
      )) %>%
      filter(Expedition %in% c("Novembro 2021")) %>%
      ggplot(aes(y = `Curated ID`, 
                 group = `Curated ID`,
                 x = Filter,
                 # x = Site,
                 fill = RRA)) +
      geom_tile() +
      # geom_text(aes(label= sprintf("%0.2f", round(RRA, digits = 2))),
      #           stat = "identity",
      #           colour = "black", size = 4) +
      facet_grid(cols = vars(Site),
                 rows = vars(Order),
                 space = "free", 
                 scales = "free",
                 drop = TRUE) +
      scale_fill_gradientn(name = "RRA (%)",
                           # colours = c("white", "yellow", "green", "darkgreen", "blue"),
                           colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF",
                                           "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
                           # colours = c("white", "#FDE725FF", "#6DCD59FF", "#35B779FF", "#3E4A89FF"),
                           # colours = c("#440154FF", "#482878FF", "#3E4A89FF", "#31688EFF",
                           #             "#26828EFF", "#1F9E89FF", "#35B779FF", "#6DCD59FF",
                           #             "#B4DE2CFF", "#FDE725FF"),
                           values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
                           breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00), # Quebras exatas
                           labels = function(x) scales::number(x, accuracy = 0.01, trim = FALSE), # Precisão exata
                           limits = c(0.01, 100),
                           na.value = "#7A0403FF",
                           trans = "log10"
      ) +
      scale_colour_gradientn(name = "RRA (%)",
                             # colours = rev(c("white", "yellow", "green", "darkgreen", "blue")),
                             colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF",
                                             "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
                             # colours = rev(c("white", "#FDE72C5FF", "#6DCD59FF", "#35B779FF", "#3E4A89FF")),
                             # colours = c("#440154FF", "#482878FF", "#3E4A89FF", "#31688EFF",
                             #             "#26828EFF", "#1F9E89FF", "#35B779FF", "#6DCD59FF",
                             #             "#B4DE2CFF", "#FDE725FF"),
                             values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
                           breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00), # Quebras exatas
                             labels = function(x) scales::number(x, accuracy = 0.01, trim = FALSE), # Precisão exata
                             limits = c(0.01, 100),
                             na.value = "#7A0403FF",
                             trans = "log10") +
      theme_minimal() +
      theme(
        panel.background = element_blank(),
        panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2)),
        strip.background = element_rect(fill = "#e4e4e4"),
        strip.text = element_text(color = "black", size = rel(1.2))) +
      theme(
        plot.title = element_text(size = 20, face = "bold"),
        # axis.text.x = element_text(size = 14, angle = 20, hjust = 1, vjust = 1),
        axis.text.x = element_text(size = 14, vjust = 1),
        axis.text.y = element_text(size = 14, face = "italic"),
        axis.title.x = element_text(size = 14, face = "bold"),
        axis.title.y = element_text(size = 14, face = "bold"),
        strip.text = element_text(size = 14, face = "bold"),
        strip.text.y = element_text(angle=0),
        legend.text = element_text(size = 13),
        legend.title = element_text(size = 14, face = "bold", margin = margin(b = 10)),
        legend.key.height = unit(3, 'cm')
      ) +
      labs(fill = "RRA%",
           x = "Sample sites",
           y = "Species")
    
    tile_mce_vs_sterivex
    
    ggsave(plot = tile_mce_vs_sterivex,
           filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
                            "tile_MCE_Sterivex_HoH", "-", Sys.Date(), ".pdf", sep = ""),
           device = "pdf", units = "cm", height = 20, width = 30, dpi = 600)
    
    # Empty versus Full (SBG) HoH suggestions
    # tile_sbg_full_vs_low <- grouped_sampled_filt_tbl %>%
    #   # filter(RRA >= 0.01) %>%
    #   filter(Filter %in% c("MCE")) %>%
    #   mutate(`Curated ID` = factor(`Curated ID`, levels = rev(spp_Levels))) %>% 
    #   mutate(Site = factor(Site)) %>%
    #   mutate(Order = factor(Order)) %>%
    #   group_by(`Curated ID`, Site, Order, Level) %>%   # Agrupa por expedição, Curated ID e Point
    #   ggplot(aes(y = `Curated ID`,
    #              group = `Curated ID`,
    #              x = Level,
    #              fill = RRA)) +
    #   geom_tile() +
    #   # geom_text(aes(label= sprintf("%0.2f", round(RRA, digits = 2))), # exibindo os valores de RRA dentro dos tiles para facilitar a discussao (opcional)
    #   # stat = "identity",
    #   # colour = "black", size = 4) +
    #   facet_grid(cols = vars(Site),
    #              rows = vars(Order),
    #              space = "free",
    #              scales = "free",
    #              drop = TRUE) +
    #         scale_fill_gradientn(name = "RRA (%)",
    #                        # colours = c("white", "yellow", "green", "darkgreen", "blue"),
    #                        colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF",
    #                                        "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
    #                        # colours = c("white", "#FDE725FF", "#6DCD59FF", "#35B779FF", "#3E4A89FF"),
    #                        # colours = c("#440154FF", "#482878FF", "#3E4A89FF", "#31688EFF",
    #                        #             "#26828EFF", "#1F9E89FF", "#35B779FF", "#6DCD59FF",
    #                        #             "#B4DE2CFF", "#FDE725FF"),
    #                        values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
    #                        breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00), # Quebras exatas
    #                        labels = function(x) scales::number(x, accuracy = 0.01, trim = FALSE), # Precisão exata
    #                        limits = c(0.01, 100),
    #                        na.value = "#7A0403FF",
    #                        trans = "log10"
    #   ) +
    #   scale_colour_gradientn(name = "RRA (%)",
    #                          # colours = rev(c("white", "yellow", "green", "darkgreen", "blue")),
    #                          colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF",
    #                                          "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
    #                          # colours = rev(c("white", "#FDE72C5FF", "#6DCD59FF", "#35B779FF", "#3E4A89FF")),
    #                          # colours = c("#440154FF", "#482878FF", "#3E4A89FF", "#31688EFF",
    #                          #             "#26828EFF", "#1F9E89FF", "#35B779FF", "#6DCD59FF",
    #                          #             "#B4DE2CFF", "#FDE725FF"),
    #                          values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
    #                        breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00), # Quebras exatas
    #                          labels = function(x) scales::number(x, accuracy = 0.01, trim = FALSE), # Precisão exata
    #                          limits = c(0.01, 100),
    #                          na.value = "#7A0403FF",
    #                          trans = "log10") +
    #   theme_minimal() +
    #   theme(panel.background = element_blank(), 
    #         panel.grid.major = element_line(color = "grey",
    #                                         size = 0.2,
    #                                         linetype = 1),
    #     axis.ticks = element_blank(),
    #     axis.text = element_text(color = "black"),
    #     axis.title = element_text(color = "black", size = rel(1.2)),
    #     legend.text = element_text(color = "black"),
    #     legend.title = element_text(color = "black", size = rel(1.2)),
    #     plot.title = element_text(color = "black", size = rel(1.5)),
    #     plot.subtitle = element_text(color = "black", size = rel(1.2)),
    #     strip.background = element_rect(fill = "#e4e4e4"),
    #     strip.text = element_text(color = "black", size = rel(1.2))) +
    #   theme(
    #     plot.title = element_text(size = 20, face = "bold"),
    #     # axis.text.x = element_text(size = 14, angle = 20, hjust = 1, vjust = 1),
    #     axis.text.x = element_text(size = 14, vjust = 1),
    #     axis.text.y = element_text(size = 14, face = "italic"),
    #     axis.title.x = element_text(size = 14, face = "bold"),
    #     axis.title.y = element_text(size = 14, face = "bold"),
    #     strip.text = element_text(size = 14, face = "bold"),
    #     strip.text.y = element_text(angle=0),
    #     legend.text = element_text(size = 13),
    #     legend.title = element_text(size = 14, face = "bold", margin = margin(b = 10)),
    #     legend.key.height = unit(3, 'cm')) +
    #   labs(fill = "RRA (%)",
    #        x = "Sample sites",
    #        y = "Species")
    #   
    # tile_sbg_full_vs_low
    # 
    #  ggsave(plot = tile_sbg_full_vs_low, 
    #        filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
    #                         "tile_sbg_Full_vs_Low_HoH", "-", Sys.Date(), ".pdf", sep = ""),
    #        device = "pdf", units = "cm", height = 20, width = 30, dpi = 600)
     
     # Tile all samples
    tile_all_samples <- grouped_sampled_filt_tbl %>%
      mutate(`Curated ID` = factor(`Curated ID`, levels = rev(spp_Levels))) %>% 
      mutate(Site = factor(Site)) %>%
      mutate(Order = factor(Order)) %>%
      group_by(`Curated ID`, Site, Order, Level) %>%   # Agrupa por expedição, Curated ID e Point
      ggplot(aes(y = `Curated ID`,
                 group = `Curated ID`,
                 x = Filter,
                 fill = RRA)) +
      geom_tile() +
      # geom_text(aes(label= sprintf("%0.2f", round(RRA, digits = 2))), # exibindo os valores de RRA dentro dos tiles para facilitar a discussao (opcional)
      # stat = "identity",
      # colour = "black", size = 4) +
      facet_grid(cols = vars(Sample),
                 rows = vars(Order),
                 space = "free",
                 scales = "free",
                 drop = TRUE) +
            scale_fill_gradientn(name = "RRA (%)",
                           # colours = c("white", "yellow", "green", "darkgreen", "blue"),
                           colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF",
                                           "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
                           # colours = c("white", "#FDE725FF", "#6DCD59FF", "#35B779FF", "#3E4A89FF"),
                           # colours = c("#440154FF", "#482878FF", "#3E4A89FF", "#31688EFF",
                           #             "#26828EFF", "#1F9E89FF", "#35B779FF", "#6DCD59FF",
                           #             "#B4DE2CFF", "#FDE725FF"),
                           values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
                           breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00), # Quebras exatas
                           labels = function(x) scales::number(x, accuracy = 0.01, trim = FALSE), # Precisão exata
                           limits = c(0.01, 100),
                           na.value = "#7A0403FF",
                           trans = "log10"
      ) +
      scale_colour_gradientn(name = "RRA (%)",
                             # colours = rev(c("white", "yellow", "green", "darkgreen", "blue")),
                             colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF",
                                             "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
                             # colours = rev(c("white", "#FDE72C5FF", "#6DCD59FF", "#35B779FF", "#3E4A89FF")),
                             # colours = c("#440154FF", "#482878FF", "#3E4A89FF", "#31688EFF",
                             #             "#26828EFF", "#1F9E89FF", "#35B779FF", "#6DCD59FF",
                             #             "#B4DE2CFF", "#FDE725FF"),
                             values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
                           breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00), # Quebras exatas
                             labels = function(x) scales::number(x, accuracy = 0.01, trim = FALSE), # Precisão exata
                             limits = c(0.01, 100),
                             na.value = "#7A0403FF",
                             trans = "log10") +
      theme_minimal() +
      theme(panel.background = element_blank(), 
            panel.grid.major = element_line(color = "grey",
                                            size = 0.2,
                                            linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2)),
        strip.background = element_rect(fill = "#e4e4e4"),
        strip.text = element_text(color = "black", size = rel(1.2))) +
      theme(
        plot.title = element_text(size = 20, face = "bold"),
        # axis.text.x = element_text(size = 14, angle = 20, hjust = 1, vjust = 1),
        axis.text.x = element_text(size = 14, vjust = 1),
        axis.text.y = element_text(size = 14, face = "italic"),
        axis.title.x = element_text(size = 14, face = "bold"),
        axis.title.y = element_text(size = 14, face = "bold"),
        strip.text = element_text(size = 14, face = "bold"),
        strip.text.y = element_text(angle=0),
        legend.text = element_text(size = 13),
        legend.title = element_text(size = 14, face = "bold", margin = margin(b = 10)),
        legend.key.height = unit(3, 'cm')) +
      labs(fill = "RRA (%)",
           x = "Sample sites",
           y = "Species")
      
    tile_all_samples
    
     ggsave(plot = tile_all_samples, 
           filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
                            "tile_all_samples", "-", Sys.Date(), ".pdf", sep = ""),
           device = "pdf", units = "cm", height = 20, width = 30, dpi = 600)
     
  }
}

# Venn diagram 
{
  # Filtering
  resume_venn <- grouped_sampled_filt_tbl %>%
    select(`Curated ID`, Genus, Family, Order, Filter, Level, Expedition, Site) %>%
    unique()
  
  # 1a. Common spp between MCE and Sterivex
  spp_comm <- resume_venn %>% 
    filter(Expedition %in% "Novembro 2021") %>% 
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(n_filtros = n_distinct(Filter)) %>%
    filter(n_filtros >= 2) %>%
    select(`Curated ID`) # Trocar para spp, genero, familia e ordem
  View(spp_comm)
  
  # 1b. Exclusive spp MCE and Sterivex
  spp_diff <- resume_venn %>%
    filter(Expedition %in% "Novembro 2021") %>% 
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(MCE_count = sum(as.integer(Filter == "MCE")),
              STX_count = sum(as.integer(Filter == "Sterivex")),
              MCE_only = MCE_count >= 1 & STX_count == 0,
              STX_only = MCE_count == 0 & STX_count >= 1) %>% 
    ungroup()
  View(spp_diff)
  
  # 2a. Common spp between Full and Low
  spp_comm_2 <- resume_venn %>% 
    filter(Filter %in% "MCE") %>%
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(n_Level = n_distinct(Level)) %>%
    filter(n_Level >= 2) %>%
    select(`Curated ID`) # Trocar para spp, genero, familia e ordem
  View(spp_comm_2)
  
  # 2b. Exclusive spp Full and Low
  spp_diff_2 <- resume_venn %>%
    filter(Filter %in% "MCE") %>%
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(Low_count = sum(as.integer(Level == "Low")),
              High_count = sum(as.integer(Level == "Full")),
              Low_only = Low_count >= 1 & High_count == 0,
              High_only = Low_count == 0 & High_count >= 1) %>% 
    ungroup()
  View(spp_diff_2)
  
    # 3a. Common spp between Sample sites
  spp_comm_3 <- resume_venn %>% 
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(n_Level = n_distinct(Site)) %>%
    filter(n_Level >= 2) %>%
    select(`Curated ID`) # Trocar para spp, genero, familia e ordem
  View(spp_comm_3)
  
  # 3b. Exclusive spp Sample sites
  spp_diff_3 <- resume_venn %>%
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(Prainha_count = sum(as.integer(Site == "Prainha")),
              Ponte_count = sum(as.integer(Site == "Ponte")),
              Barragem_count = sum(as.integer(Site == "Barragem")),
              Fundacao_count = sum(as.integer(Site == "Fundação")),
              Prainha_only = Prainha_count >= 1 & Ponte_count == 0 & Barragem_count == 0 & Fundacao_count == 0,
              Ponte_only = Ponte_count >= 1 & Prainha_count == 0 & Barragem_count == 0 & Fundacao_count == 0,
              Barragem_only = Barragem_count >= 1 & Prainha_count == 0 & Ponte_count == 0 & Fundacao_count == 0,
              Fundacao_only = Fundacao_count >= 1 & Prainha_count == 0 & Ponte_count == 0 & Barragem_count == 0) %>% 
    ungroup()
  View(spp_diff_3)
}
```

## Analysis 

```{r, eval=FALSE,echo=FALSE}
## NMDS ----
# Converter planilha de identificações por ponto para o formato Amostras X IDs
{
  # Função necessária para juntar as abds da tabela em IDs
  sum_uniq <- function(vec) {
    if (is.character(vec)) {
      suniq <- BiocGenerics::unique(vec)
    }
    if (is.numeric(vec)) {
      suniq <- sum(vec)
    }
    return(suniq)
  }
  
  # Número de linhas (IDs diferentes) por amostra
  grouped_sampled_filt_tbl$Sample %>% table()
  
  grouped_sampled_filt_tbl %>% colnames()
  
  # FINAL_tbl_IDs <- grouped_by_ID_BLASTid %>%
  FINAL_tbl_IDs <- grouped_sampled_filt_tbl %>%
    # filter(Expedition %in% c("Novembro 2021")) %>% 
    # filter(RRA >= 0.05) %>% # definicao de qual threshold de abundancia sera usado
    # filter(Site %in% c("Ponte", "Fundacao")) %>% 
    # filter(!Sample %in% c("L2_dez20")) %>% # amostra problematica
    select(c("Sample", "Curated ID", "Year", "Level", "Site", "Filter", "RRA", "Expedition")) %>% 
    pivot_wider(id_cols = c("Sample", "Year", "Level", "Site", "Filter", "Expedition"),
                names_from = `Curated ID`,
                values_from = `RRA`,
                values_fn = sum_uniq,
                names_sort = TRUE,
                names_prefix = "ID_") %>% 
    relocate(c("Sample", "Year", "Level", "Filter", "Site", "Expedition", starts_with("ID_"))) %>%  
    mutate(across(starts_with("ID_"), \(x) replace_na(x, 0))) %>% 
    mutate("Site" = as.factor(Site)) %>% 
    mutate("Level" = as.factor(Level)) 
  
  # Verificando a soma das linhas
  FINAL_tbl_IDs %>% select(starts_with(match = "ID_")) %>% rowSums(na.rm = TRUE)
}

#  Rodando o NMDS

# Preparar dados para entrada no pacote vegan
colnames(FINAL_tbl_IDs)

FINAL_tbl_IDs$`Sample` %>% unique() %>% sort()

all_IDs_NMDS_tbl <- FINAL_tbl_IDs %>% 
  mutate("Sample number" = 0) %>% 
  relocate("Sample number" )

# Associar números de amostra aos nomes de amostra
for (sample in 1:nrow(all_IDs_NMDS_tbl)) {
  all_IDs_NMDS_tbl$`Sample number`[sample] <- sample
}

# Ordenar dataframe usado no NMDS
all_IDs_NMDS_df <- all_IDs_NMDS_tbl %>%
  # mutate(Level = ifelse(Level == "Full", "Full", "Empty")) %>%
  as.data.frame()

# Nomear linhas como números de amostra
row.names(all_IDs_NMDS_df) <- all_IDs_NMDS_df$`Sample number`
all_IDs_NMDS_df <- all_IDs_NMDS_df 

# Corrigir nomes das espécies para evitar problemas na plotagem
colnames(all_IDs_NMDS_df)
colnames(all_IDs_NMDS_df)[8:ncol(all_IDs_NMDS_df)] <- colnames(all_IDs_NMDS_df)[8:ncol(all_IDs_NMDS_df)] %>%
  str_replace_all(pattern = " ", replacement = "_") %>% 
  str_replace_all(pattern = "\\.", replacement = "") %>% 
  str_replace_all(pattern = "\\(", replacement = "") %>% 
  str_replace_all(pattern = "\\)", replacement = "")

# Executando o NMDS
# Esta e a funcao que faz o NMDS. Para ela fornece-se apenas
# as colunas relativas as especies nas amostras
all_ps_vegan_ord_meta <- metaMDS(veg = all_IDs_NMDS_df[,8:ncol(all_IDs_NMDS_df)],
                                 comm = all_IDs_NMDS_df[8:ncol(all_IDs_NMDS_df)],
                                 # distance = "bray", # Leva em conta o RRA
                                 distance = "jaccard", # Considera apenas presença/ausenciaa
                                 halfchange = TRUE 
)

dim(all_IDs_NMDS_df)

# Fazer o fit das variáveis ambientais

# meta.envfit <- envfit(all_ps_vegan_ord_meta, all_IDs_NMDS_df[, c("Level", "Sample")],
meta.envfit <- envfit(all_ps_vegan_ord_meta, all_IDs_NMDS_df[, c("Filter", "Sample")],
                      permutations = 999,
                      na.rm=TRUE) # this fits environmental vectors

# Espécies 

# Fazer o fit das espécies para identificar a significância delas na explicação dos agrupamentos. 
# Esse é o passo que mais demora quando com muitas amostras e espécies.
meta.spp.fit <- envfit(all_ps_vegan_ord_meta, 
                       all_IDs_NMDS_df[,7:ncol(all_IDs_NMDS_df)], 
                       permutations = 999) # this fits species vectors

# Obter valores de p para as espécies
sps_pvals <- tibble("IDs" = names(meta.spp.fit$vectors$pvals),
                    "p-value" = meta.spp.fit$vectors$pvals)

spp.scrs <- as.data.frame(scores(meta.spp.fit, display = "vectors")) %>%
  mutate("IDs" = rownames(.)) %>%
  left_join(y = sps_pvals, by = "IDs")

{
  spp.scrs$IDs <- gsub("ID_", "", spp.scrs$IDs)
  spp.scrs$IDs <- gsub("_", " ", spp.scrs$IDs)
  } 

# Selecionar espécies significativas
# sig.spp.scrs <- spp.scrs %>%
  # filter(`p-value` <=
           # 0.05) # definir p-value aqui!

# Pontos amostrais

# Definir os valores de NMDS1 e NMDS2, e os metadados de cada amostra/ponto amostral
site.scrs <- as.data.frame(scores(all_ps_vegan_ord_meta, display = "sites")) %>%
  mutate("Sample number" = as.double(row.names(.))) %>%
  left_join(y = all_IDs_NMDS_df[, c("Sample number",
                                    "Sample",
                                    "Level",
                                    "Site",
                                    "Filter"
  )], 
  by = "Sample number")

# Determinar centroides
scrs <- scores(all_ps_vegan_ord_meta, display = "sites")

cent <- aggregate(scrs ~ Filter, data = site.scrs, FUN = "mean")

# Calcular elipses
NMDS <- data.frame("MDS1" = all_ps_vegan_ord_meta$points[, 1],
                   "MDS2" = all_ps_vegan_ord_meta$points[, 2],
                   "Filter" = as.factor(all_IDs_NMDS_df$Filter), check.names = FALSE)

NMDS.mean <- aggregate(NMDS[, 1:2], list(group = NMDS$Filter), "mean")


# Elipses
{
  plot(all_ps_vegan_ord_meta)
  
  # Sobrepor as elipses
  ord <- ordiellipse(ord = all_ps_vegan_ord_meta, 
                     groups = all_IDs_NMDS_df$Filter,
                     display = "sites",
                     kind = "ehull", conf = 0.95, label = T)
  
  
  # Funcao do vegan de calcular elipses
  veganCovEllipse<-function (cov, center = c(0, 0), scale = 1, npoints = 100) 
  {
    theta <- (0:npoints) * 2 * pi/npoints
    Circle <- cbind(cos(theta), sin(theta))
    t(center + scale * t(Circle %*% chol(cov)))
  }
  
  df_ell <- data.frame()
  
  for(g in levels(NMDS$Filter)){
    print(g)
    df_ell <- 
      rbind(df_ell, 
            cbind(as.data.frame(with(NMDS[NMDS$Filter==g,],
                                     veganCovEllipse(
                                       ord[[g]]$cov,
                                       ord[[g]]$center,
                                       ord[[g]]$scale))),
                  Filter=g))}
}

# Configurar plot NMDS

# Definir paleta de cores
{
  # paletteer::paletteer_d("colorBlindness::SteppedSequential5Steps")
  # paletteer::paletteer_d("colorBlindness::SteppedSequential5Steps")[1:3]
  # paletteer::paletteer_d("colorBlindness::SteppedSequential5Steps")[c(1:3, 2,7,9,10)]
  # 
  # my_cols <- c("BAM" = "#0F6B99FF",
  #              "PARAC" = "#6B990FFF",
  #              "SAM" = "#99540FFF",
  #              "SFC" = "#A3CC51FF",
  #              "SFI" = "#7EC3E5FF",
  #              "SFM" = "#E5B17EFF") 
  
  cores <- c("#f1c232",
             "#3381b1")
  
}

# Plot
{
  NMDS_MCE_Sterivex <-
    ggplot(data = site.scrs,
           aes(x=NMDS1, 
               y=NMDS2)) +
    
    # Elipses 
    ggforce::geom_mark_ellipse(inherit.aes = FALSE,
                               data = df_ell,
                               aes(x = NMDS1,
                                   y = NMDS2,
                                   group = Filter,
                                   label = Filter,
                                   col = Filter,
                                   fill = Filter
                               ),
                               alpha=0.10,
                               # n = 200,
                               linetype=2,
                               expand = 0,
                               label.fontsize = 14,
                               con.cap = 0.1
    ) +
    # Niveis amostrais
    geom_point(aes(x=NMDS1,
                   y=NMDS2,
                   fill = Filter,
                   # label = `Sampling unit`,  #descomentar se quiser exibir os nomes dos `Sampling sites`s
                   col=Filter,
                   group = Filter,
                   shape = Site),
               stroke = 0.5,
               alpha = 0.75,
               size = 3) +
    
    # Nomes dos `Sampling sites`s amostrais
    ggrepel::geom_text_repel(aes(label = Site),  #descomentar bloco se quiser exibir os nomes dos pontos
                             # hjust=0.5,
                             # vjust=2.75,
                             size=4.5,
                             direction = "both",
                             segment.size = 0.25,
                             segment.alpha=0.1,
                             # min.segment.length = 1.5,
                             force = 3,
                             max.overlaps = 100,
                             fontface = "bold") +
    
    # Vetores das IDs
    geom_segment(data = sig.spp.scrs, aes(x = 0,
                                          xend = NMDS1,
                                          y = 0,
                                          yend = NMDS2),
                 arrow = arrow(length = unit(0.1, "cm")),
                 colour = "grey30",
                 alpha = 0.5,
                 lwd = 0.3) + #add vector arrows of significant species
    
    # Nomes das IDs
    ggrepel::geom_text_repel(data = sig.spp.scrs,
                             aes(x=NMDS1, y=NMDS2, label = IDs),
                             size = 3.5,
                             alpha = 0.75,
                             direction = "both",
                             segment.size = 0.25,
                             segment.alpha = 0.1,
                             max.overlaps = 100) +
    
    # Centroides
    geom_point(data = cent,
               aes(x = NMDS1,
                   y = NMDS2, # colour = Filter,
                   fill = Filter
               ),
               size = 8,
               colour = "#222222",
               alpha = 0.75,
               shape  = 13
    ) + 
    coord_fixed(expand = c(0.5))+
    theme_light() +
    theme(
      panel.background = element_blank(),
      # panel.grid.major = element_blank(),
      panel.grid.major = element_line(color = "grey",
                                      size = 0.2,
                                      linewidth = 1),
      # axis.ticks = element_line(color = "grey"),
      axis.ticks = element_blank(),
      axis.text = element_text(color = "black"),
      axis.title = element_text(color = "black", size = rel(1.2)),
      # legend.background = element_blank(),
      # legend.key = element_blank(),
      legend.text = element_text(color = "black"),
      legend.title = element_text(color = "black", size = rel(1.2)),
      plot.title = element_text(color = "black", size = rel(1.5)),
      plot.subtitle = element_text(color = "black", size = rel(1.2))) +
    theme(plot.title = element_text(size = 14, face = "bold"),
          axis.text.x = element_text(size = 12),
          axis.text.y = element_text(size = 12),
          axis.title.x = element_text(size = 13, face = "bold"),
          axis.title.y = element_text(size = 13, face = "bold"),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 13),
          # legend.position = "bottom"
    ) +
    labs(x = "NMDS1",
         y = "NMDS2",
         title = "Composição da ictiofauna",
         # subtitle = "Filtro MCE versus Sterivex") +
         subtitle = "Stress: ",format(round(all_ps_vegan_ord_meta$stress))) +
    scale_fill_manual(values = cores) +
    scale_colour_manual(values = cores) 
  # + scale_fill_manual(values = viridis::turbo(n = 5)) # cores das formas das samples/pontos
  
  
  NMDS_MCE_Sterivex
  
  
  # Salvar em pdf 
  # ggsave(plot = NMDS_MCE_Sterivex, 
  #        filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
  #                         "NDMS_MCE_STX_filt", "-", Sys.Date(), ".pdf", sep = ""),
  #        device = "pdf",
  #        units = "cm",
  #        height = 25,
  #        width = 25,
  #        dpi = 600)
}

## PCoA ----

# Dados
## Usando a tabela all_IDs_NMDS_df gerada para o NDMDS

## Executando o PCoA para MCE versus Sterivex
{
  # Filtrando os dados

mce_stx_IDs_NMDS_df <- all_IDs_NMDS_df %>% 
  filter(Expedition %in% c("Novembro 2021")) %>% 
  select(where(~ any(. != 0))) 

colnames(mce_stx_IDs_NMDS_df[,8:ncol(mce_stx_IDs_NMDS_df)])

#Criando a Matriz de distancia
pcOa_dist <- vegan::vegdist(x = mce_stx_IDs_NMDS_df[,8:ncol(mce_stx_IDs_NMDS_df)],
                            method = "jaccard",
                            binary = TRUE)

pcOa <- cmdscale(pcOa_dist, eig = TRUE)

# Porcentagens

percent_pcOa <- pcOa$eig / sum(pcOa$eig) * 100
percent_pcOa[1:2]

ordiplot(pcOa, display = 'sites', type = 'text')

# Espécies 

# Fazer o fit das espécies para identificar a significância delas na explicação dos agrupamentos. 
# Esse é o passo que mais demora quando com muitas amostras e espécies.
meta.spp.fit <- envfit(pcOa, 
                       mce_stx_IDs_NMDS_df[,8:ncol(mce_stx_IDs_NMDS_df)], 
                       permutations = 999) # this fits species vectors

# Obter valores de p para as espécies
sps_pvals <- tibble("IDs" = names(meta.spp.fit$vectors$pvals),
                    "p-value" = meta.spp.fit$vectors$pvals)

spp.scrs <- as.data.frame(scores(meta.spp.fit, display = "vectors")) %>%
  mutate("IDs" = rownames(.)) %>%
  left_join(y = sps_pvals, by = "IDs")

{
  spp.scrs$IDs <- gsub("ID_", "", spp.scrs$IDs)
  spp.scrs$IDs <- gsub("_", " ", spp.scrs$IDs)
  } 

# Selecionar espécies significativas
sig.spp.scrs <- spp.scrs

t_spp <- spp.scrs %>%
  filter(`p-value` <=
           0.05) # definir p-value aqui!

# Pontos amostrais

#Definir os valores de PCoA1 e PCoA2, e os metadados de cada amostra/ponto amostral
site.scrs <- as.data.frame(scores(pcOa, display = "sites"))

colnames(site.scrs) <- c("PCoA1", "PCoA2")

site.scrs <- site.scrs %>% 
  mutate("Sample number" = as.double(row.names(.))) %>%
  left_join(y = mce_stx_IDs_NMDS_df[, c("Sample number",
                                    "Sample",
                                    "Level",
                                    "Site",
                                    "Filter",
                                    "Expedition"
  )],
  by = "Sample number")

# Determinar centroides
scrs <- scores(pcOa, display = "sites")

cent <- aggregate(scrs ~ Filter, data = site.scrs, FUN = "mean")

# Calcular elipses
PCoA <- data.frame("PCoA1" = pcOa$points[, 1],
                   "PCoA2" = pcOa$points[, 2],
                   "Filter" = as.factor(mce_stx_IDs_NMDS_df$Filter), check.names = FALSE)

PCoA.mean <- aggregate(PCoA[, 1:2], list(group = PCoA$Filter), "mean")

# Elipses

# plot(all_ps_vegan_ord_meta)
ordiplot(pcOa, display = 'sites', type = 'text')

# Sobrepor as elipses
ord <- ordiellipse(ord = pcOa, 
                   groups = mce_stx_IDs_NMDS_df$Filter,
                   display = "sites",
                   kind = "ehull", conf = 0.95, label = T)


# Funcao do vegan de calcular elipses
veganCovEllipse <-function (cov, center = c(0, 0), scale = 1, npoints = 100) 
{
  theta <- (0:npoints) * 2 * pi/npoints
  Circle <- cbind(cos(theta), sin(theta))
  t(center + scale * t(Circle %*% chol(cov)))
}

df_ell <- data.frame()

for(g in levels(PCoA$Filter)){
  print(g)
  df_ell <- 
    rbind(df_ell, 
          cbind(as.data.frame(with(PCoA[PCoA$Filter==g,],
                                   veganCovEllipse(
                                     ord[[g]]$cov,
                                     ord[[g]]$center,
                                     ord[[g]]$scale))),
                Filter=g))}

# Configurar plot PCoA

# Definir paleta de cores

cores <- c("#f1c232",
           "#3381b1")
# Plot

  PCoA_MCE_Sterivex <-
    ggplot(data = site.scrs,
           aes(x=PCoA1, 
               y=PCoA2)) +
    
    # Elipses 
    ggforce::geom_mark_ellipse(inherit.aes = FALSE,
                               data = df_ell,
                               aes(x = Dim1,
                                   y = Dim2,
                                   group = Filter,
                                   label = Filter,
                                   col = Filter,
                                   fill = Filter),
                               alpha=0.10,
                               # n = 200,
                               linetype=2,
                               expand = 0,
                               label.fontsize = 14,
                               con.cap = 0.1
    ) +
    # Niveis amostrais
    geom_point(aes(x=PCoA1,
                   y=PCoA2,
                   fill = Filter,
                   # label = `Sampling unit`,  #descomentar se quiser exibir os nomes dos `Sampling sites`s
                   col=Filter,
                   group = Filter,
                   shape = Site),
               stroke = 0.5,
               alpha = 0.75,
               size = 3) +
    
  # Nomes dos Sampling sites amostrais
    ggrepel::geom_text_repel(aes(label = Site),  #descomentar bloco se quiser exibir os nomes dos pontos
                             # hjust=0.5,
                             # vjust=2.75,
                             size=4.5,
                             direction = "both",
                             segment.size = 0.25,
                             segment.alpha=0.1,
                             # min.segment.length = 1.5,
                             force = 3,
                             max.overlaps = 100,
                             fontface = "bold") +

  # Vetores das IDs
  # geom_segment(data = sig.spp.scrs, aes(x = 0,
  #                                       xend = Dim1,
  #                                       y = 0,
  #                                       yend = Dim2),
  #              arrow = arrow(length = unit(0.1, "cm")),
  #              colour = "grey30",
  #              alpha = 0.5,
  #              lwd = 0.3) + #add vector arrows of significant species

  # Nomes das IDs
  # ggrepel::geom_text_repel(data = sig.spp.scrs,
  #                          aes(x=Dim1, y=Dim2, label = IDs),
  #                          size = 3.5,
  #                          alpha = 0.75,
  #                          direction = "both",
  #                          segment.size = 0.25,
  #                          segment.alpha = 0.1,
  #                          max.overlaps = 25) +
  
  # Centroides
  geom_point(data = cent,
             aes(x = Dim1,
                 y = Dim2, # colour = Filter,
                 fill = Filter),
             size = 8,
             colour = "#222222",
             alpha = 0.75,
             shape  = 13) + 
    coord_fixed(expand = c(0.5))+
    theme_light() +
    theme(
      panel.background = element_blank(),
      # panel.grid.major = element_blank(),
      panel.grid.major = element_line(color = "grey",
                                      linewidth = 0.2,
                                      linetype = 1),
      # axis.ticks = element_line(color = "grey"),
      axis.ticks = element_blank(),
      axis.text = element_text(color = "black"),
      axis.title = element_text(color = "black", size = rel(1.2)),
      # legend.background = element_blank(),
      # legend.key = element_blank(),
      legend.text = element_text(color = "black"),
      legend.title = element_text(color = "black", size = rel(1.2)),
      plot.title = element_text(color = "black", size = rel(1.5)),
      plot.subtitle = element_text(color = "black", size = rel(1.2))) +
    theme(plot.title = element_text(size = 14, face = "bold"),
          axis.text.x = element_text(size = 12),
          axis.text.y = element_text(size = 12),
          axis.title.x = element_text(size = 13, face = "bold"),
          axis.title.y = element_text(size = 13, face = "bold"),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 13),
          # legend.position = "bottom"
    ) +
    labs(x = "PCoA1 (37.30%)",
         y = "PCoA2 (32.43%)",
         # title = "Composição da ictiofauna",
         # subtitle = "Filtro MCE versus Sterivex") +
         # subtitle = "Stress: ",format(round(all_ps_vegan_ord_meta$stress))
    ) +
    scale_fill_manual(values = cores) +
    scale_colour_manual(values = cores) 
  # + scale_fill_manual(values = viridis::turbo(n = 5)) # cores das formas das samples/pontos
  
  PCoA_MCE_Sterivex

# Salvar em pdf 
ggsave(plot = PCoA_MCE_Sterivex,
       filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
                        "PCoA_MCE_Sterivex_clean",
                        # "PCoA_MCE_Sterivex_full",
                        "-", Sys.Date(), ".pdf", sep = ""),
       device = "pdf",
       units = "cm",
       height = 25,
       width = 25,
       dpi = 600)
}

## Executando o PCoA para Full versus Low
{
  # Filtrando os dados

mce_IDs_NMDS_df <- all_IDs_NMDS_df %>% 
  filter(Filter %in% c("MCE")) %>% 
  select(where(~ any(. != 0))) 

colnames(mce_IDs_NMDS_df[,8:ncol(mce_IDs_NMDS_df)])

#Criando a Matriz de distancia
pcOa_dist <- vegan::vegdist(x = mce_IDs_NMDS_df[,8:ncol(mce_IDs_NMDS_df)],
                            method = "jaccard",
                            binary = TRUE)

pcOa <- cmdscale(pcOa_dist, eig = TRUE)

ordiplot(pcOa, display = 'sites', type = 'text')

# Porcentagens

percent_pcOa <- pcOa$eig / sum(pcOa$eig) * 100
percent_pcOa[1:2]

# Espécies 

# Fazer o fit das espécies para identificar a significância delas na explicação dos agrupamentos. 
# Esse é o passo que mais demora quando com muitas amostras e espécies.
meta.spp.fit <- envfit(pcOa, 
                       mce_IDs_NMDS_df[,8:ncol(mce_IDs_NMDS_df)], 
                       permutations = 999) # this fits species vectors

# Obter valores de p para as espécies
sps_pvals <- tibble("IDs" = names(meta.spp.fit$vectors$pvals),
                    "p-value" = meta.spp.fit$vectors$pvals)

spp.scrs <- as.data.frame(scores(meta.spp.fit, display = "vectors")) %>%
  mutate("IDs" = rownames(.)) %>%
  left_join(y = sps_pvals, by = "IDs")

{
  spp.scrs$IDs <- gsub("ID_", "", spp.scrs$IDs)
  spp.scrs$IDs <- gsub("_", " ", spp.scrs$IDs)
  } 

# Selecionar espécies significativas
sig.spp.scrs <- spp.scrs 

t.spp.scrs <- spp.scrs %>%
  filter(`p-value` <=
           0.05) # definir p-value aqui!

# Pontos amostrais

#Definir os valores de PCoA1 e PCoA2, e os metadados de cada amostra/ponto amostral
site.scrs <- as.data.frame(scores(pcOa, display = "sites"))

colnames(site.scrs) <- c("PCoA1", "PCoA2")

site.scrs <- site.scrs %>% 
  mutate("Sample number" = as.double(row.names(.))) %>%
  left_join(y = mce_IDs_NMDS_df[, c("Sample number",
                                    "Sample",
                                    "Level",
                                    "Site",
                                    # "Filter",
                                    "Expedition"
  )],
  by = "Sample number")

# Determinar centroides
scrs <- scores(pcOa, display = "sites")

cent <- aggregate(scrs ~ Level, data = site.scrs, FUN = "mean")

# Calcular elipses
PCoA <- data.frame("PCoA1" = pcOa$points[, 1],
                   "PCoA2" = pcOa$points[, 2],
                   # "Filter" = as.factor(mce_IDs_NMDS_df$Filter), check.names = FALSE)
                   "Level" = as.factor(mce_IDs_NMDS_df$Level), check.names = FALSE)

PCoA.mean <- aggregate(PCoA[, 1:2], list(group = PCoA$Level), "mean")

# Elipses

# plot(all_ps_vegan_ord_meta)
ordiplot(pcOa, display = 'sites', type = 'text')

# Sobrepor as elipses
ord <- ordiellipse(ord = pcOa, 
                   groups = mce_IDs_NMDS_df$Level,
                   display = "sites",
                   kind = "ehull", conf = 0.95, label = T)


# Funcao do vegan de calcular elipses
veganCovEllipse <-function (cov, center = c(0, 0), scale = 1, npoints = 100) 
{
  theta <- (0:npoints) * 2 * pi/npoints
  Circle <- cbind(cos(theta), sin(theta))
  t(center + scale * t(Circle %*% chol(cov)))
}

df_ell <- data.frame()

for(g in levels(PCoA$Level)){
  print(g)
  df_ell <- 
    rbind(df_ell, 
          cbind(as.data.frame(with(PCoA[PCoA$Level==g,],
                                   veganCovEllipse(
                                     ord[[g]]$cov,
                                     ord[[g]]$center,
                                     ord[[g]]$scale))),
                Level=g))}

# Configurar plot PCoA

# Definir paleta de cores

cores <- c("#3381b1",
           "#f1c232")

# Plot

  PCoA_MCE <-
    ggplot(data = site.scrs,
           aes(x=PCoA1, 
               y=PCoA2)) +
    
    # Elipses 
    ggforce::geom_mark_ellipse(inherit.aes = FALSE,
                               data = df_ell,
                               aes(x = Dim1,
                                   y = Dim2,
                                   group = Level,
                                   label = Level,
                                   col = Level,
                                   fill = Level
                               ),
                               alpha=0.10,
                               # n = 200,
                               linetype=2,
                               expand = 0,
                               label.fontsize = 14,
                               con.cap = 0.1
    ) +
    # Niveis amostrais
    geom_point(aes(x=PCoA1,
                   y=PCoA2,
                   fill = Level,
                   # label = `Sampling unit`,  #descomentar se quiser exibir os nomes dos `Sampling sites`s
                   col=Level,
                   group = Level,
                   shape = Site),
               stroke = 0.5,
               alpha = 0.75,
               size = 3) +
    
    # Nomes dos Sampling sites amostrais
    ggrepel::geom_text_repel(aes(label = Site),  #descomentar bloco se quiser exibir os nomes dos pontos
                             # hjust=0.5,
                             # vjust=2.75,
                             size=4.5,
                             direction = "both",
                             segment.size = 0.25,
                             segment.alpha=0.1,
                             # min.segment.length = 1.5,
                             force = 3,
                             max.overlaps = 100,
                             fontface = "bold") +
    
    # Vetores das IDs
    geom_segment(data = sig.spp.scrs, aes(x = 0,
                                          xend = Dim1,
                                          y = 0,
                                          yend = Dim2),
                 arrow = arrow(length = unit(0.1, "cm")),
                 colour = "grey30",
                 alpha = 0.5,
                 lwd = 0.3) + #add vector arrows of significant species
    
    # Nomes das IDs
    ggrepel::geom_text_repel(data = sig.spp.scrs,
                             aes(x=Dim1, y=Dim2, label = IDs),
                             size = 3.5,
                             alpha = 0.75,
                             direction = "both",
                             segment.size = 0.25,
                             segment.alpha = 0.1,
                             max.overlaps = 25) +
    
    # Centroides
    geom_point(data = cent,
               aes(x = Dim1,
                   y = Dim2, # colour = Filter,
                   fill = Level
               ),
               size = 8,
               colour = "#222222",
               alpha = 0.75,
               shape  = 13
    ) + 
    coord_fixed(expand = c(0.5))+
    theme_light() +
    theme(
      panel.background = element_blank(),
      # panel.grid.major = element_blank(),
      panel.grid.major = element_line(color = "grey",
                                      size = 0.2,
                                      linetype = 1),
      # axis.ticks = element_line(color = "grey"),
      axis.ticks = element_blank(),
      axis.text = element_text(color = "black"),
      axis.title = element_text(color = "black", size = rel(1.2)),
      # legend.background = element_blank(),
      # legend.key = element_blank(),
      legend.text = element_text(color = "black"),
      legend.title = element_text(color = "black", size = rel(1.2)),
      plot.title = element_text(color = "black", size = rel(1.5)),
      plot.subtitle = element_text(color = "black", size = rel(1.2))) +
    theme(plot.title = element_text(size = 14, face = "bold"),
          axis.text.x = element_text(size = 12),
          axis.text.y = element_text(size = 12),
          axis.title.x = element_text(size = 13, face = "bold"),
          axis.title.y = element_text(size = 13, face = "bold"),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 13),
          # legend.position = "bottom"
    ) +
    labs(x = "PCoA1 (29.67%)",
         y = "PCoA2 (22.33%)",
         # title = "Composição da ictiofauna",
         # subtitle = "Filtro MCE versus Sterivex") +
         # subtitle = "Stress: ",format(round(all_ps_vegan_ord_meta$stress))
    ) +
    scale_fill_manual(values = cores) +
    scale_colour_manual(values = cores) 
  # + scale_fill_manual(values = viridis::turbo(n = 5)) # cores das formas das samples/pontos
  
  PCoA_MCE

# Salvar em pdf 
ggsave(plot = PCoA_MCE, 
       filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
                        "PCoA_MCE",
                        # "PCoA_MCE_Sterivex_full",
                        "-", Sys.Date(), ".pdf", sep = ""),
       device = "pdf",
       units = "cm",
       height = 25,
       width = 25,
       dpi = 600)
}

## Species accumulation curves ----
  
## Definindo os dados 
wider_grouped_sampled_filt_ASVs <- grouped_sampled_filt_tbl %>%
  select(c(Sample,
           Site,
           Level,
           `Curated ID`,
           ASVs,
           Expedition,
           Filter,
           Level)) %>%
  group_by(Sample, `Curated ID`, Reads) %>%
  unique %>%
  pivot_wider(names_from = `Curated ID`,
              values_from = ASVs,
              names_sort = TRUE) %>%
  as.data.frame() %>%
  mutate(across(everything(), ~replace_na(.x, 0))) %>%
  column_to_rownames("Sample")

accum_ASVs <- specaccum(wider_grouped_sampled_filt_ASVs[5:ncol(wider_grouped_sampled_filt_ASVs)],
                             method = "rarefaction"
                             # method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
plot(accum_ASVs$individuals, accum_ASVs$richness)

## Getting into a dataframe
accum_ASVs_df <- tibble("Sites" = c(0, accum_ASVs$sites),
                        "Individuals" = c(0, accum_ASVs$individuals),
                        "Richness" = c(0, accum_ASVs$richness),
                        "sd" = c(0, accum_ASVs$sd))
## PLotting
accum_ASVs_df %>% 
ggplot(aes(x = Individuals,
           y = Richness)) +
    geom_line(linewidth = 2.5,
              alpha = 0.50) +
    geom_point(size = 1) +
    # geom_errorbar(aes(ymin = Richness - sd, ymax = Richness + sd), width = 0.01) +
    geom_ribbon(aes(ymin = Richness - sd, ymax = Richness + sd), 
                alpha = 0.15,
                size = 0.1) +
    theme_minimal() +
    theme(panel.grid.major = element_line(color = "grey", size = 0.4),
          panel.grid.minor = element_line(color = "lightgrey", size = 0.1)) +
    labs(x = "Species richness", 
         y = "Number of ASVs",
         title = "Collector's curve MCE and Sterivex samples")

# All samples (Richness vs Samples)
{
  ## Definindo os dados 
  wider_grouped_sampled_filt <- grouped_sampled_filt_tbl %>%
    select(c(Sample,
             Site,
             Level,
             `Curated ID`,
             Reads,
             Expedition,
             Filter,
             Level)) %>%
    group_by(Sample, `Curated ID`, Reads) %>%
    unique %>%
    pivot_wider(names_from = `Curated ID`,
                values_from = Reads,
                names_sort = TRUE) %>% 
    as.data.frame() %>% 
    mutate(across(everything(), ~replace_na(.x, 0))) %>%
    column_to_rownames("Sample")
  
    ## Function to transform values into 1 and zeros
  jaccarize <- function(x) {
    # Using ifelse to check each element of the vector x
    ifelse(x == 0, 0, 1)
    }

  wider_grouped_sampled_filt <- wider_grouped_sampled_filt %>%
    mutate(across(c(5:ncol(wider_grouped_sampled_filt)), jaccarize))
  
  ## Combined MCE and STX
  wider_grouped_sampled_filt_stx_mce <- wider_grouped_sampled_filt %>%
    filter(Expedition %in% "Novembro 2021")

  accum_stx_mce <- specaccum(wider_grouped_sampled_filt_stx_mce[5:ncol(wider_grouped_sampled_filt_stx_mce)],
                             # method = "rarefaction"
                             method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_stx_mce$sites, accum_stx_mce$richness)

  ## Getting into a dataframe
  accum_mce_stx_df <- tibble("Sites" = c(0, accum_stx_mce$sites),
                             "Individuals" = c(0, accum_stx_mce$individuals),
                             "Richness" = c(0, accum_stx_mce$richness),
                             "sd" = c(0, accum_stx_mce$sd),
                             "Filter" = "MCE and Sterivex (November 2021)")
  
  ## MCE only
  wider_grouped_sampled_filt_mce <- wider_grouped_sampled_filt %>%
    filter(Filter %in% "MCE")

  accum_mce <- specaccum(wider_grouped_sampled_filt_mce[5:ncol(wider_grouped_sampled_filt_mce)],
                             # method = "rarefaction"
                             method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_mce$sites, accum_mce$richness)
  
  ## Getting into a dataframe
  accum_mce_df <- tibble("Sites" = c(0, accum_mce$sites),
                         "Individuals" = c(0, accum_mce$individuals),
                         "Richness" = c(0, accum_mce$richness),
                         "sd" = c(0, accum_mce$sd),
                         "Filter" = "MCE")
  
  ## MCE (November 2021) only
  wider_grouped_sampled_filt_mce21 <- wider_grouped_sampled_filt %>%
    filter(Filter %in% "MCE") %>% 
    filter(Expedition %in% "Novembro 2021")  

  accum_mce21 <- specaccum(wider_grouped_sampled_filt_mce21[5:ncol(wider_grouped_sampled_filt_mce21)],
                             # method = "rarefaction"
                             method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_mce21$sites, accum_mce21$richness)
  
  ## Getting into a dataframe
  accum_mce21_df <- tibble("Sites" = c(0, accum_mce21$sites),
                         "Individuals" = c(0, accum_mce21$individuals),
                         "Richness" = c(0, accum_mce21$richness),
                         "sd" = c(0, accum_mce21$sd),
                         "Filter" = "MCE (November 2021)")
  
  ## Sterivex only
  wider_grouped_sampled_filt_stx <- wider_grouped_sampled_filt %>%
    filter(Filter %in% "Sterivex")

  accum_stx <- specaccum(wider_grouped_sampled_filt_stx[5:ncol(wider_grouped_sampled_filt_stx)],
                             # method = "rarefaction"
                             method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_stx$sites, accum_stx$richness)

  ## Getting into a dataframe
  accum_stx_df <- tibble("Sites" = c(0, accum_stx$sites),
                         "Individuals" = c(0, accum_stx$individuals),
                         "Richness" = c(0, accum_stx$richness),
                         "sd" = c(0, accum_stx$sd),
                         "Filter" = "Sterivex")
  
  ## MCE and Sterivex 
  accum_all <- specaccum(wider_grouped_sampled_filt[5:ncol(wider_grouped_sampled_filt)],
                             # method = "rarefaction"
                             method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_all$sites, accum_all$richness)

  ## Getting into a dataframe
  accum_all_df <- tibble("Sites" = c(0, accum_all$sites),
                         "Individuals" = c(0, accum_all$individuals),
                         "Richness" = c(0, accum_all$richness),
                         "sd" = c(0, accum_all$sd),
                         "Filter" = "All samples")
  
  ## Combining dataframes
  combined_accum <- bind_rows(#accum_mce_stx_df, 
                              # accum_mce_df,
                              accum_mce21_df, 
                              accum_stx_df
                              # accum_all_df, 
                              )
  
  combined_accum$Filter <- factor(combined_accum$Filter,
                                  levels = c("All samples",
                                             "MCE",
                                             "MCE and Sterivex (November 2021)",
                                             "MCE (November 2021)",
                                             "Sterivex"))
  
  ## Plotting
  collector_plot_all <- 
    combined_accum %>% 
    ggplot(aes(x = Sites,
               y = Richness,
               color = Filter,
               fill = Filter)) +
    geom_line(linewidth = 2.5,
              alpha = 0.50) +
    geom_point(size = 1) +
    # geom_errorbar(aes(ymin = Richness - sd, ymax = Richness + sd), width = 0.01) +
    geom_ribbon(aes(ymin = Richness - sd, ymax = Richness + sd), 
                alpha = 0.15,
                size = 0.1) +
    scale_color_manual(values = c(# "#9C27B0",
                                  # "#4CAF50", 
                                  "#F57F17", 
                                  "#2196F3"
                                  # "#F44336" 
                                  )) +  
    scale_fill_manual(values = c(# "#673AB7",
                                 # "#005602",
                                 "#F57F55",
                                 "#0D47A1"
                                 # "#800020"
                                 )) +
    theme_minimal() +
    theme(panel.grid.major = element_line(color = "grey", size = 0.4),
          panel.grid.minor = element_line(color = "lightgrey", size = 0.1)) +
    labs(x = "Number of samples", 
         y = "Species richness",
         # title = "Collector's curve MCE and Sterivex samples"
         )
  
  collector_plot_all
  
  ggsave(file = paste0(figs_path,"/", "collector_plot_STX_MCE21-", Sys.Date(), ".pdf", collapse = ""),
         plot = collector_plot_all,
         device = "pdf",
         width = 16,
         height = 12,
         units = "cm",
         dpi = 300)
    }

# All samples (Richness vs ASVs)
{
  ## Definindo os dados 
  wider_grouped_sampled_filt_ASVs
  
  ## Combined MCE and STX
  wider_grouped_sampled_filt_ASVs_mce_stx <- wider_grouped_sampled_filt_ASVs %>%
    filter(Expedition %in% "Novembro 2021")

  accum_stx_mce <- specaccum(wider_grouped_sampled_filt_ASVs_mce_stx[5:ncol(wider_grouped_sampled_filt_ASVs_mce_stx)],
                             method = "rarefaction"
                             # method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_stx_mce$individuals, accum_stx_mce$richness)

  ## Getting into a dataframe
  accum_mce_stx_df <- tibble("Sites" = c(0, accum_stx_mce$sites),
                             "Individuals" = c(0, accum_stx_mce$individuals),
                             "Richness" = c(0, accum_stx_mce$richness),
                             "sd" = c(0, accum_stx_mce$sd),
                             "Filter" = "MCE and Sterivex (November 2021)")
  
  ## MCE only
  wider_grouped_sampled_filt_ASVs_mce <- wider_grouped_sampled_filt_ASVs %>%
    filter(Filter %in% "MCE")

  accum_mce <- specaccum(wider_grouped_sampled_filt_ASVs_mce[5:ncol(wider_grouped_sampled_filt_ASVs_mce)],
                             method = "rarefaction"
                             # method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_mce$sites, accum_mce$richness)
  
  ## Getting into a dataframe
  accum_mce_df <- tibble("Sites" = c(0, accum_mce$sites),
                         "Individuals" = c(0, accum_mce$individuals),
                         "Richness" = c(0, accum_mce$richness),
                         "sd" = c(0, accum_mce$sd),
                         "Filter" = "MCE")
  
  ## MCE (November 2021) only
  wider_grouped_sampled_filt_ASVs_mce21 <- wider_grouped_sampled_filt_ASVs %>%
    filter(Filter %in% "MCE") %>% 
    filter(Expedition %in% "Novembro 2021")  

  accum_mce21 <- specaccum(wider_grouped_sampled_filt_ASVs_mce21[5:ncol(wider_grouped_sampled_filt_ASVs_mce21)],
                             method = "rarefaction"
                             # method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_mce21$individuals, accum_mce21$richness)
  
  ## Getting into a dataframe
  accum_mce21_df <- tibble("Sites" = c(0, accum_mce21$sites),
                         "Individuals" = c(0, accum_mce21$individuals),
                         "Richness" = c(0, accum_mce21$richness),
                         "sd" = c(0, accum_mce21$sd),
                         "Filter" = "MCE (November 2021)")
  
  ## Sterivex only
  wider_grouped_sampled_filt_ASVs_stx <- wider_grouped_sampled_filt_ASVs %>%
    filter(Filter %in% "Sterivex")

  accum_stx <- specaccum(wider_grouped_sampled_filt_stx[5:ncol(wider_grouped_sampled_filt_stx)],
                             method = "rarefaction"
                             # method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_stx$individuals, accum_stx$richness)

  ## Getting into a dataframe
  accum_stx_df <- tibble("Sites" = c(0, accum_stx$sites),
                         "Individuals" = c(0, accum_stx$individuals),
                         "Richness" = c(0, accum_stx$richness),
                         "sd" = c(0, accum_stx$sd),
                         "Filter" = "Sterivex")
  
  ## MCE and Sterivex 
  accum_all <- specaccum(wider_grouped_sampled_filt_ASVs[5:ncol(wider_grouped_sampled_filt_ASVs)],
                             method = "rarefaction"
                             # method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_all$individuals, accum_all$richness)

  ## Getting into a dataframe
  accum_all_df <- tibble("Sites" = c(0, accum_all$sites),
                         "Individuals" = c(0, accum_all$individuals),
                         "Richness" = c(0, accum_all$richness),
                         "sd" = c(0, accum_all$sd),
                         "Filter" = "All samples")
  
  ## Combining dataframes
  combined_accum <- bind_rows(accum_mce_stx_df, accum_mce_df, accum_mce21_df, accum_stx_df, accum_all_df, )
  
  combined_accum$Filter <- factor(combined_accum$Filter,
                                  levels = c("All samples",
                                             "MCE",
                                             "MCE and Sterivex (November 2021)",
                                             "MCE (November 2021)",
                                             "Sterivex"))
  
  ## Plotting
  collector_plot_all_ASVs <- 
    combined_accum %>% 
    ggplot(aes(x = Individuals,
               y = Richness,
               color = Filter,
               fill = Filter)) +
    geom_line(linewidth = 2.5,
              alpha = 0.50) +
    geom_point(size = 1) +
    # geom_errorbar(aes(ymin = Richness - sd, ymax = Richness + sd), width = 0.01) +
    geom_ribbon(aes(ymin = Richness - sd, ymax = Richness + sd), 
                alpha = 0.15,
                size = 0.1) +
    scale_color_manual(values = c("#9C27B0",
                                  "#4CAF50", 
                                  "#F57F17", 
                                  "#2196F3",
                                  "#F44336" 
                                  )) +  
    scale_fill_manual(values = c("#673AB7",
                                 "#005602",
                                 "#F57F17",
                                 "#0D47A1",
                                 "#800020"
                                 )) +
    theme_minimal() +
    theme(panel.grid.major = element_line(color = "grey", size = 0.4),
          panel.grid.minor = element_line(color = "lightgrey", size = 0.1)) +
    labs(x = "Number of ASVs",
         y = "Species richness",
         title = "Collector's curve MCE and Sterivex samples (ASVs)") 
  
  collector_plot_all_ASVs
  
  ggsave(file = paste0(figs_path,"/", "collector_plot_STX-", Sys.Date(), ".pdf", collapse = ""),
         plot = collector_plot_STX,
         device = "pdf",
         width = 16,
         height = 12,
         units = "cm",
         dpi = 300)
    }

## Diversity indexes ----

grouped_sampled_filt_tbl %>% colnames()

# Creating "OTU" table ----

gsf_phyloseq <- grouped_sampled_filt_tbl %>%
  pivot_wider(id_cols = c("Curated ID"),
              values_from = RRA,
              values_fn = sum_uniq,
              names_from = "Sample",
              names_sort = TRUE
              ) %>%
  mutate(across(everything(), replace_na, replace = 0)) %>% 
  # mutate_if(is.numeric, round) %>%
  mutate_if(is.numeric, jaccarize) %>%
  as.data.frame() %>%
  `rownames<-`(.$`Curated ID`) %>% 
  select(-c("Curated ID")) %>% 
  filter(rowSums(.) != 0) %>% # Remover linhas cuja soma é zero
  select(which(!colSums(., na.rm=TRUE) %in% 0)) # Remover colunas cuja soma é zero
 
dim(gsf_phyloseq)

gsf_phyloseq %>% rowSums() ==0

# Creating taxonomy table ----

grouped_sampled_filt_tbl %>% colnames()

blast_tax_table <- grouped_sampled_filt_tbl %>%
  select(c("Curated ID",
           "Genus",
           "Family",
           "Order"
           )) %>% 
  unique() %>% 
  as.data.frame() %>% 
  `rownames<-`(.$`Curated ID`) %>% 
  select(-c("Curated ID")) %>% 
  as.matrix()

# Creating sample metadata table ----

metadata_tbl <- grouped_sampled_filt_tbl %>%
  select(c("Sample", "Site", "Year", "Expedition", "Level", "Filter")) %>% 
  unique() %>%
  as.data.frame() %>% 
  `rownames<-`(.$`Sample`)

# Creating final phyloseq object ----

final_ps <- phyloseq::phyloseq(otu_table(gsf_phyloseq, taxa_are_rows = T),
                               sample_names(metadata_tbl),
                               tax_table(blast_tax_table))

# Estimating richness (final_ps)

div_est_phylo <- estimate_richness(final_ps,
                                   split = T,
                                   measures = c("Observed", "Chao1", "Obs./Chao1", "se.chao1", 
                                                "ACE", "se.ACE", "Shannon","Simpson", "InvSimpson")) %>%
  as_tibble(rownames = "Sample") %>% 
  mutate("Obs./Chao1" = Observed/Chao1) %>%
  pivot_longer(cols = c("Observed", "Chao1", "Obs./Chao1", "se.chao1",
                        "ACE", "se.ACE", "Shannon","Simpson", "InvSimpson"),
               names_to = "Indice",
               values_to = "Values") %>%
  mutate(Indice = factor(Indice, levels = c("Observed", "Chao1", "Obs./Chao1", "se.chao1",
                                            "ACE", "se.ACE", "Shannon","Simpson", "InvSimpson"))) %>% 
  left_join(y = metadata_tbl,
            by = "Sample")

 ## Sterivex versus MCE (All samples)
 div_stx_mce_plot <-
   div_est_phylo %>%
     # filter(Expedition %in% c("Novembro 2021")) %>%
   filter(Indice %in% c("Observed",
                        "Chao1",
                        "Obs./Chao1",
                        "Shannon",
                        "Simpson")) %>%
   ggplot(aes(y = Values,
              x = Filter ,
              fill = Filter,
              alpha = 0.1)) +
   geom_boxplot(col = "#282828",
                alpha = 1,
                linewidth = 0.3,
                outlier.shape = NA) +
   geom_jitter(aes(col = Site,
                   shape = Level),
               alpha = 0.75,
               height = 0,
               width = 0.25) +
   facet_wrap(nrow = 1,
              ~Indice,
              scales = "free") +
   theme_bw(base_size = 14) +
   theme(axis.text.x = element_text(angle = 45,hjust = 1,vjust = 1),
         axis.title = element_text(size = 14),
         strip.text = element_text(size = 14)) +
   scale_fill_manual(values = c("#f8e199", "#99c0d8")) +
   scale_shape_manual(values = c("Full" = 16, "Low" = 4)) +
   guides(fill = "none") +
   labs(x = "Filter type",
        y = "Value")
 
 div_stx_mce_plot
 
 ## MCE (2020, 2021, 2022)
 div_all_plot <-
   div_est_phylo %>%
     # filter(Filter %in% c("MCE")) %>%
   filter(Indice %in% c("Observed",
                          "Chao1",
                          "Obs./Chao1",
                          "Shannon",
                          "Simpson")) %>%
   ggplot(aes(y = Values,
              x = Level ,
              groups = Level,
              fill = Level,
              alpha = 0.1)) +
   geom_boxplot(col = "#282828",
                alpha = 1,
                linewidth = 0.3,
                outlier.shape = NA) +
   geom_jitter(aes(col = Site,
                   shape = Filter),
               alpha = 1,
               height = 0,
               width = 0.25) +
   # geom_signif(
   #   comparisons = list(c("Full", "Low")),
   #   map_signif_level = F,
   #   col = "#111111",
   #   test = "wilcox.test",
   #   test.args = list(exact = FALSE)
   #   ) +
   facet_wrap(nrow = 1,
              ~Indice,
              scales = "free") +
   theme_bw(base_size = 14) +
   theme(axis.text.x = element_text(angle = 45,hjust = 1,vjust = 1),
         axis.title = element_text(size = 14),
         strip.text = element_text(size = 14)) +
   scale_fill_manual(values = c("#99c0d8","#f8e199")) +
   scale_shape_manual(values = c(1, 3)) +
   guides(fill = "none") +
   xlab(label = "Level") +
   ylab(label = "Value") 

 div_all_plot
 
 # Combining diversity plots in one figure
 
 ggarrange(div_stx_mce_plot, div_all_plot,
                    labels = c("A", "B"),
                    ncol = 2, nrow = 1)
 
```
