---
title: "LI paper scripts"
author: 
  - "Mendes, GA"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    code_download: yes
    df_print: paged
    keep_md: yes
    theme: flatly
    toc: true
    toc_depth: 5
    toc_float: true
  pdf_document: 
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}

# load project environment and Rmd necessary libraries
# load(file = "/home/gabriel/projetos/LI_paper/Environments/...RData")

```


**This pipeline integrates public tools available for metabarcoding analyses. To share or reproduce this content, please require authors' consent.**  
**Contact:** gabrielmendesbrt@gmail.com 

# Bioinformatics

## Load  R libs

```{r, eval=FALSE,echo=FALSE}
# Loading libraries ----
{ library("tidyverse")
  library("Biostrings")
  library("DECIPHER")
  library("phyloseq")
  library("vegan")
  library("readxl")
  library("ggpubr")
  library("ggplot2")
  library("ggsignif")
  library("ggrepel")
  library("openxlsx")
  }
```

## Create project folders

```{r, eval=FALSE,echo=FALSE}
# Creating and loading paths ----
{
  prjct_path <- "/home/gabriel/projetos/LI_paper"
  prjct_radical <- "eDNA_Lagoa-dos-Ingleses"
  scripts_path <- paste0(prjct_path,"/scripts")
  results_path <- paste0(prjct_path,"/results")
  figs_path <- paste0(results_path,"/figures")
  tbl_path <- paste0(results_path,"/tables")
  env_path <- paste0(prjct_path,"/environment")
  paths <- c(scripts_path, results_path, figs_path, tbl_path, env_path)
}

for (dir in paths) {
  if (dir.exists(dir)) {
    print(paste("The directory", dir, "already exists!"))
  } else {
    print(paste("Making", dir, "!" ))
    dir.create(dir) 
  }
}
```

## Data acquiring

```{r, eval=FALSE,echo=FALSE}
pre_raw_results_tbl <- read_excel("/home/gabriel/projetos/lagoa_ingleses/tabelas/curated/curated-Complete_analysis_results-2024-01-10.xlsx") %>% tibble()
curated_ids_tbl <- read_excel("/home/gabriel/projetos/lagoa_ingleses/tabelas/curated/curated_lagoa_ingleses-ASVs_x_amostras-2024-01-09.xlsx") %>% tibble()
blast_tax <- read.csv("/home/gabriel/projetos/lagoa_ingleses/tabelas/raw/tax_blast.csv", sep = ",", check.names = FALSE)
```

## Data merging and transformation 

```{r, eval=FALSE,echo=FALSE}
## Selecting colunms ----

# Curated IDs

curated_ids_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()

curated_ids_tbl <-
  curated_ids_tbl %>%
  select(c( "ASV header", "ASV (Sequence)", "Curated ID", "Final ID (BLASTn)", "BLASTn pseudo-score", 
            "Class (BLASTn)", "Curated Order (BLASTn)")) %>%
  unique()

# View(curated_ids_tbl)

# Complete analysis

# pre_raw_results_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()

pre_raw_results_tbl <-
  pre_raw_results_tbl %>%
  select(c("Sample", "Unique_File_name", "OTU", "Abundance", "Obs. Curadoria", "Possible contamination", 
           "Read origin", "Primer expected length", "ASV Size (pb)", "ASV header")) %>%
  mutate(Sample = str_replace(Sample, "__", "_"))

# View(pre_raw_results_tbl)

## Integrating curated identifications and reordering ----

# pre_raw_results_tbl %>% colnames() %>% paste0(collapse = '",\n"') %>% cat()

raw_results_tbl <- pre_raw_results_tbl %>%
  left_join(curated_ids_tbl,
            by = c("ASV header")) %>%
  select(c("ASV header", #oriundas da curated_ids_tbl
           "ASV (Sequence)",
           "Curated ID",
           "Final ID (BLASTn)",
           "BLASTn pseudo-score",
           "Class (BLASTn)",
           "Curated Order (BLASTn)",
           "Sample", #oriundas da pre_raw_samples_tbl
           "Unique_File_name",
           "OTU",
           "Abundance",
           "Obs. Curadoria",
           "Read origin",
           "Primer expected length",
           "ASV Size (pb)",
           "ASV header"
  ))

# View(raw_results_tbl)

## Adding BLASTn's complete taxonomy ----

blast_tax <- blast_tax[2:85] %>% 
  rename(`ASV (Sequence)` = "ASV")

blast_tax_less <- blast_tax %>%
  left_join(raw_results_tbl,
           by = "ASV (Sequence)") %>% 
  mutate("Genus (BLASTn)" = str_split_fixed(string = .$`Curated ID`, 
                                           pattern = " ",
                                           n = 2)[,1]) %>% 
  select(c("ASV (Sequence)",
           # "Order (BLASTn)",
           "Family (BLASTn)",
           "Genus (BLASTn)"))
         
raw_results_tbl <- raw_results_tbl %>%
  left_join(blast_tax_less,
            by = "ASV (Sequence)") %>% 
  select(c("ASV header", "ASV (Sequence)", "Curated ID", "Final ID (BLASTn)", "BLASTn pseudo-score", 
           "Class (BLASTn)", "Family (BLASTn)", "Genus (BLASTn)", "Curated Order (BLASTn)", "Sample", 
           "Unique_File_name", "OTU", "Abundance", "Obs. Curadoria", "Read origin", "Primer expected length", 
           "ASV Size (pb)")) %>% 
  unique() %>% 
  rename("Order (BLASTn)" = "Curated Order (BLASTn)")

# Correcting family IDs february 10th 2025

raw_results_tbl <- raw_results_tbl %>% 
  mutate(`Family (BLASTn)` = 
           ifelse(`Genus (BLASTn)` %in% c("Psalidodon",
                                          "Moenkhausia",
                                          "Astyanax",
                                          "Hemigrammus",
                                          "Oligosarcus"), 
                  "Acestrorhamphidae",
                  `Family (BLASTn)`)) %>% 
  mutate(`Family (BLASTn)` = ifelse(`Genus (BLASTn)` %in% "Piabina",
                                    "Stevardiidae",
                                    `Family (BLASTn)`))

# Correcting Hoplias IDs february 25th 2025

raw_results_tbl <- raw_results_tbl %>%
  mutate(`Curated ID` =
           ifelse(`Curated ID` %in% c("Hoplias malabaricus",
                                      "Hoplias intermedius"),
                  "Hoplias sp.",
                  `Curated ID`))

# View(raw_results_tbl)

## 4o Preliminary filtering ----

# Filtering out of range reads, NA identifications, Mammalian and Aves reads
# with BLASTn pseudo-score lower than 98 and reads from negative controls

pre_filt_tbl <-
  raw_results_tbl %>%
  filter(!`Primer expected length` %in% "out of range") %>% 
  filter(!`Curated ID` == "NA") %>% 
  filter(!(`Class (BLASTn)` %in% c("Mammalia", "Aves") & `BLASTn pseudo-score` < 98)) %>% 
  filter(!(Sample %in% c("EM113_NEGPCR1", "EM135c4c5_NEGPCR1", "EM135c4c5_NEGPCR2", "EM149_NEGPCR1", 
                         "EM149_NEGPCR2", "EM156_NegPCR1", "S724_NEGPCR1", "br_jan_22", "br_nov21"))) %>% 
  # Filter(!(`Possible contamination` %in% c("Possible contamination"))) %>% 
  filter(!(`Obs. Curadoria` %in% c("Possible contamination"))) 

# View(pre_filt_tbl)

## 5o Adding metadata from sampling 

# pre_filt_tbl$Sample %>% unique() %>% sort() %>% paste0(collapse = '",\n"') %>% cat()
{
  # sample
  Sample <- c("L1_nov_dec_20_mi", "L2_nov20", "L2_dez20", "L1_out21", "L2_out21", "L3_out21", 
              "L4_out21", "L1_nov21", "L2_nov21", "L3_nov21", "L4_nov21", "STX_L1_nov21", 
              "STX_L2_nov21", "STX_L3_nov21", "STX_L4_nov21", "L1_jan22", "L2_jan22", "L3_jan22", 
              "L4_jan22")
  
  # Expedition
  Expedition <- c("Novembro e Dezembro 2020", "Novembro 2020", "Dezembro 2020", "Outubro 2021", 
                  "Outubro 2021", "Outubro 2021", "Outubro 2021", "Novembro 2021", "Novembro 2021", 
                  "Novembro 2021", "Novembro 2021", "Novembro 2021", "Novembro 2021", "Novembro 2021", 
                  "Novembro 2021", "Janeiro 2022", "Janeiro 2022", "Janeiro 2022", "Janeiro 2022")
  
  # Year
  Year <- c("2020", "2020", "2020", "2021", "2021", "2021", "2021", "2021", "2021", "2021", "2021", 
            "2021", "2021", "2021", "2021", "2022", "2022", "2022", "2022")
  
  # Site
  # Site <- c("Fundação", "Ponte", "Ponte", "Prainha", "Barragem", "Ponte", "Fundação", "Prainha", 
  #           "Barragem", "Ponte", "Fundação", "Prainha", "Barragem", "Ponte", "Fundação", "Prainha", 
  #           "Barragem", "Ponte", "Fundação")  

  # Site
  Site <- c("P1", "P2", "P2", "P3", "P4", "P2", "P1", "P3", "P4", "P2", "P1",
  "P3", "P4", "P2", "P1", "P3", "P4", "P2", "P1")
  
  # Filter
  Filter <- c("MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "MCE", "Sterivex", 
              "Sterivex", "Sterivex", "Sterivex", "MCE", "MCE", "MCE", "MCE")
  
  # Run
  Run <- c("Run2_ago21", "Run4_out21", "Run4_out21", "Run5_dez21", "Run5_dez21", "Run5_dez21", 
           "Run5_dez21", "Run5_dez21", "Run5_dez21", "Run5_dez21", "Run5_dez21", "EM156", "EM156", 
           "EM156", "EM156", "EM156", "EM156", "EM156", "EM156")

  # Level
  Level <- c("High", "High", "High", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", "Low", 
             "Low", "Low", "Low", "High", "High", "High", "High")
}

# As tibble
metadata_tbl <- tibble(Run, Sample, Filter, Site, Expedition, Year, Level)

curated_full_tbl <- left_join(pre_filt_tbl, metadata_tbl,
                              by = "Sample") %>%
  select(c("ASV header", "ASV (Sequence)", "Curated ID", "Final ID (BLASTn)", "BLASTn pseudo-score", 
           "Class (BLASTn)", "Family (BLASTn)", "Genus (BLASTn)", "Order (BLASTn)", "Sample", 
           "Unique_File_name", "OTU", "Abundance", "Read origin", "Primer expected length", "ASV Size (pb)", 
           "Run", "Sample", "Filter", "Site", "Expedition", "Year", "Level")) %>% 
  filter(Filter %in% "MCE")

View(curated_full_tbl)
```

## Suplementary tables
```{r, eval=FALSE,echo=FALSE}
## Raw table ----

# Creating a table to evaluate the raw data after pipeline processing

raw_resume <- raw_results_tbl %>% 
  group_by(Sample) %>%
  mutate("Abd total" = sum(Abundance)) %>%
  ungroup() %>%
  group_by(`Curated ID`) %>% 
  mutate("Abundancia total" = sum(Abundance)) %>% 
  ungroup %>% 
  group_by(`Final ID (BLASTn)`) %>% 
  mutate("Reads totais" = sum(Abundance)) %>% 
  ungroup() %>% 
  group_by(Sample) %>%
  mutate("RRA by sample" = (Abundance/`Abd total`)*100) %>% 
  reframe(
    "Total reads" = `Reads totais`,
    "Total abundance" = `Abundancia total`,
    "RRA" = sum(`RRA by sample`),
    "ASVs" = length(unique(`ASV header`)),
    "OTUs" = length(unique(OTU)),
    "Reads" = sum(Abundance),
    "Class" = `Class (BLASTn)`,
    "Order" = `Order (BLASTn)`,
    "Family" = `Family (BLASTn)`,
    "Genus" = `Genus (BLASTn)`,
    "Curated ID" = `Curated ID`
  ) %>%
  unique()

## Suplementar table A ----

# Creating a table to use to evaluate the unfiltered data

# 1st Longer table
dt_all_resume <- curated_full_tbl %>% 
  group_by(Level, Site, Filter, Year) %>%
  mutate("Abd total" = sum(Abundance)) %>%
  ungroup() %>%
  group_by(`Curated ID`) %>% 
  mutate("Abundancia total" = sum(Abundance)) %>% 
  ungroup %>% 
  group_by(`Final ID (BLASTn)`) %>% 
  mutate("Reads totais" = sum(Abundance)) %>% 
  ungroup() %>% 
  group_by(`Curated ID`, `Final ID (BLASTn)`, `Site`, 
           `Level`, `Filter`, `Year`) %>%
  mutate("RRA no periodo" = (Abundance/`Abd total`)*100) %>% 
  reframe(
    "Total reads" = `Reads totais`,
    "Total abundance" = `Abundancia total`,
    "RRA" = sum(`RRA no periodo`),
    "ASVs" = length(unique(`ASV header`)),
    "OTUs" = length(unique(OTU)),
    "Reads" = sum(Abundance),
    "Class" = `Class (BLASTn)`,
    "Order" = `Order (BLASTn)`,
    "Family" = `Family (BLASTn)`,
    "Genus" = `Genus (BLASTn)`
  ) %>%
  unique() 

# View(dt_all_resume)

# Testing if RRA was calculated right
dt_all_resume %>%
  group_by( `Site`, `Level`, `Filter`, `Year`) %>%
  summarise(total_RRA = sum(RRA, na.rm = TRUE)) %>%
  ungroup() %>% View()

# Wider table to evaluate unfiltered data
{wider_dt_all_resume <- dt_all_resume %>% 
  mutate(RRA = round(RRA,digits = 4)) %>% 
  ungroup() %>% 
  unite(Site, Level, Year, Filter, col= "Site_Level_Year") %>% 
  pivot_wider(id_cols = c("Class","Curated ID", "Final ID (BLASTn)", "Total reads", "Total abundance"),
              # names_from = Site_Level_Year_Filter,
              names_from = Site_Level_Year,
              values_from =  c("RRA","ASVs","OTUs","Reads"),
              names_glue = "{Site_Level_Year}_{.value}") %>%
  select(sort(colnames(.))) %>% 
  relocate("Class","Curated ID","Final ID (BLASTn)","Total reads", "Total abundance",
           #2020 MCE
           "P1_High_2020_MCE_Reads",
           "P1_High_2020_MCE_ASVs",
           "P1_High_2020_MCE_OTUs",
           "P1_High_2020_MCE_RRA",
           "P2_High_2020_MCE_Reads",
           "P2_High_2020_MCE_ASVs",
           "P2_High_2020_MCE_OTUs",
           "P2_High_2020_MCE_RRA",
           #2021 MCE
           "P1_Low_2021_MCE_Reads",
           "P1_Low_2021_MCE_ASVs",
           "P1_Low_2021_MCE_OTUs",
           "P1_Low_2021_MCE_RRA",
           "P2_Low_2021_MCE_Reads",
           "P2_Low_2021_MCE_ASVs",
           "P2_Low_2021_MCE_OTUs",
           "P2_Low_2021_MCE_RRA",
           "P3_Low_2021_MCE_Reads",
           "P3_Low_2021_MCE_ASVs",
           "P3_Low_2021_MCE_OTUs",
           "P3_Low_2021_MCE_RRA",
           "P4_Low_2021_MCE_Reads",
           "P4_Low_2021_MCE_ASVs",
           "P4_Low_2021_MCE_OTUs",
           "P4_Low_2021_MCE_RRA",
           #2022 MCE
           "P1_High_2022_MCE_Reads",
           "P1_High_2022_MCE_ASVs",
           "P1_High_2022_MCE_OTUs",
           "P1_High_2022_MCE_RRA",
           "P2_High_2022_MCE_Reads",
           "P2_High_2022_MCE_ASVs",
           "P2_High_2022_MCE_OTUs",
           "P2_High_2022_MCE_RRA",
           "P3_High_2022_MCE_Reads",
           "P3_High_2022_MCE_ASVs",
           "P3_High_2022_MCE_OTUs",
           "P3_High_2022_MCE_RRA",
           "P4_High_2022_MCE_Reads",
           "P4_High_2022_MCE_ASVs",
           "P4_High_2022_MCE_OTUs",
           "P4_High_2022_MCE_RRA"
  )}

View(wider_dt_all_resume)

write.csv(wider_dt_all_resume, paste0(tbl_path, "/wider_dt_all_resume_2025.csv"))

## Suplementar table B ----

# Creating the table with data to be used in analysis
# Filtered data, only Actinopteri and reads >=100

filt_tbl <- curated_full_tbl %>% 
  filter(`Class (BLASTn)` %in% "Actinopteri",
         !`Curated ID` %in% c("Salmo salar")) %>% 
  group_by(`Curated ID`, Site, Level, Filter, Year) %>% 
  mutate(Reads = sum(Abundance)) %>% 
  filter(Reads >= 100) %>% 
  ungroup()

View(filt_tbl)

# Filtered and grouped data by Samples
grouped_sampled_filt_tbl <- curated_full_tbl %>% 
  filter(`Class (BLASTn)` %in% "Actinopteri",
         !`Curated ID` %in% c("Salmo salar")) %>% 
  group_by(`Curated ID`, Sample) %>% 
  mutate(Reads = sum(Abundance)) %>% 
  filter(Reads >= 100) %>% 
  ungroup() %>% 
  # creating total abundance by sample
  group_by(Sample) %>%
  mutate("Abd total" = sum(Abundance)) %>%
  ungroup() %>% 
  # calculating RRA by Sample
  group_by(`Curated ID`, Sample) %>%
  mutate("RRA na amostra" = (Abundance/`Abd total`*100)) %>% 
  summarize("Site" = dplyr::first(Site),
          "Year" = dplyr::first(Year),
          "Expedition" = dplyr::first(Expedition),
          "Level" = dplyr::first(Level),
          "Filter" = dplyr::first(Filter),
          "Class" = dplyr::first(`Class (BLASTn)`),
          "Order" = dplyr::first(`Order (BLASTn)`),
          "Family" = dplyr::first(`Family (BLASTn)`),
          "Genus" = dplyr::first(`Genus (BLASTn)`),
          "Reads" = sum(Abundance),
          "Sample abundance" = dplyr::first(`Abd total`),
          "RRA" = sum(`RRA na amostra`),
          "ASVs" = length(unique(`ASV header`)),
          "OTUs" = length(unique(OTU)),
          ) 

View(grouped_sampled_filt_tbl)

# Is RRA correct?
grouped_sampled_filt_tbl %>%
  group_by(Sample) %>%
  summarise(total_RRA = sum(RRA, na.rm = TRUE)) %>%
  ungroup()

# Filtered and grouped data by Year, Site and Filter

# A tabela wider precisa dessas informacoes separadamente
grouped_filt_tbl <- filt_tbl %>%
  # creating total abundance by period
  group_by(Level, Site, Filter, Year) %>%
  mutate("Abd total" = sum(Abundance)) %>%
  ungroup() %>%
  # creating total abundance by Curated ID
  group_by(`Curated ID`) %>% 
  mutate("Abundancia total" = sum(Abundance)) %>% 
  ungroup() %>% 
  # calculating RRA by period
  group_by(`Curated ID`, Site, Level, Filter, Year) %>%
  mutate("RRA no periodo" = (Abundance/`Abd total`)*100) %>% 
  reframe("Curated ID" = `Curated ID`,
          "Class" = `Class (BLASTn)`,
          "Order" = `Order (BLASTn)`,
          "Family" = `Family (BLASTn)`,
          "Genus" = `Genus (BLASTn)`,
          "Reads" = sum(Abundance),
          "Period abundance" = `Abd total`,
          "Total abundance" = `Abundancia total`, 
          "RRA" = sum(`RRA no periodo`),
          "ASVs" = length(unique(`ASV header`)),
          "OTUs" = length(unique(OTU))) %>% 
  unique()

View(grouped_filt_tbl)

# Is RRA correct?
grouped_filt_tbl %>%
  group_by(Site, Level, Filter, Year) %>%
  summarise(total_RRA = sum(RRA, na.rm = TRUE)) %>%
  ungroup()

# Tabela wider com os dados filtrados
  wider_filt_tbl <- grouped_filt_tbl %>% 
  mutate(RRA = round(RRA,digits = 4)) %>% 
  ungroup() %>% 
  unite(Site, Level, Year, col= "Site_Level_Year") %>% 
  pivot_wider(id_cols = c("Curated ID", "Total abundance"),
              names_from = Site_Level_Year,
              values_from =  c("RRA","ASVs","OTUs","Reads"),
              names_glue = "{Site_Level_Year}_{.value}") %>%
  select(sort(colnames(.))) %>% 
  relocate("Curated ID",
           "Total abundance",
           #2020 MCE
           "P1_High_2020_Reads",
           "P1_High_2020_ASVs",
           "P1_High_2020_OTUs",
           "P1_High_2020_RRA",
           "P2_High_2020_Reads",
           "P2_High_2020_ASVs",
           "P2_High_2020_OTUs",
           "P2_High_2020_RRA",
           #2021 MCE
           "P1_Low_2021_Reads",
           "P1_Low_2021_ASVs",
           "P1_Low_2021_OTUs",
           "P1_Low_2021_RRA",
           "P2_Low_2021_Reads",
           "P2_Low_2021_ASVs",
           "P2_Low_2021_OTUs",
           "P2_Low_2021_RRA",
           "P3_Low_2021_Reads",
           "P3_Low_2021_ASVs",
           "P3_Low_2021_OTUs",
           "P3_Low_2021_RRA",
           "P4_Low_2021_Reads",
           "P4_Low_2021_ASVs",
           "P4_Low_2021_OTUs",
           "P4_Low_2021_RRA",
           #2022 MCE
           "P3_High_2022_Reads",
           "P3_High_2022_ASVs",
           "P3_High_2022_OTUs",
           "P3_High_2022_RRA",
           "P4_High_2022_Reads",
           "P4_High_2022_ASVs",
           "P4_High_2022_OTUs",
           "P4_High_2022_RRA")

View(wider_filt_tbl)

write.csv(wider_filt_tbl, paste0(tbl_path, "/wider_dt_filt_2025.csv"))
```


## Basic stats

```{r, eval=FALSE,echo=FALSE}
options(scipen = 999)

# Reads by sample after sequencing ----

## Raw stats

# Numero de reads por etapa

seq_read_counts <- read_excel("/home/gabriel/projetos/peixes-eDNA/analises/dez_23/runs_2_4_5_EM156/results/LI-reads_and_seqs_counts2024-01-08.xlsx") %>% 
  tibble() %>%
  # escolher apenas amostras relacionadas ao filtro MCE
  filter(Sample %in% c(
                        "L1_nov_dec_20_mi", "L2_nov20", "L2_dez20", "L1_out21", "L2_out21", "L3_out21", "L4_out21", "L1_nov21", "L2_nov21", "L3_nov21",
                        # "L4_nov21", "STX__L1_nov21", "STX__L2_nov21", "STX__L3_nov21", "STX__L4_nov21", "br_nov21", 
                        "L1_jan22", "L2_jan22", "L3_jan22", "L4_jan22", "br_jan_22"
                        ,"EM113_NEGPCR1", "EM135c4c5_NEGPCR1", "EM135c4c5_NEGPCR2", "EM149_NEGPCR1", "EM149_NEGPCR2", "EM156_NegPCR1", "S724_NEGPCR1"
  )) %>% 
  summarise("Total reads" = sum(`Raw reads (pairs)`))    
             
# Reads post-pipeline and filters

# Obtendo total de reads por amostra apos pipeline e filtragens, que foram usadas nas analises 
# Filtros: Sem NA, apenas Actinopteri, Curated ID >= 100 reads, 

# filt_raw_reads <- grouped_filt_tbl %>%
filt_raw_reads <- grouped_sampled_filt_tbl %>%
  # group_by(`Curated ID`) %>% # grouped_filt_tbl
  group_by(Sample) %>% # grouped_filt_tbl
  summarise(total_abd_filt = sum(Reads))

View(filt_raw_reads)

sum(filt_raw_reads$total_abd_filt)

# Total reads in ASVs, OTUs, Orders, Families, Genus, Species and Nspecies for raw data

raw_stats <- raw_resume %>%
  group_by(Class) %>% 
  reframe(
    "Total reads" = sum(Reads),
    "ASVs" = sum(ASVs),
    "OTUs" = sum(OTUs),
    "Orders" = length(unique(Order)),
    "Families" = length(unique(Family)),
    "Genera" = length(unique(Genus)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & !grepl("sp\\.", `Curated ID`)])),
    "Nspecies" = length(unique(`Curated ID`)) - Species
  ) %>%
  unique()

View(raw_stats)

# Total reads in ASVs, OTUs, Orders, Families, Genus, Species and Nspecies

# Obtendo infos sobre as reads que foram usadas nas analises
# Quantas ASVs, OTUs, Ordens, Familias, Generos, Especies e IDs a nivel de Genus spp.
# Escolher qual agrupamento: Classe, Ordem, Familia, Genero ou Especie

grouped_filt_stats <- grouped_filt_tbl %>% #dependendo da tabela, resultados diferentes
  group_by(Class) %>% # Grouped by Class
  # group_by(Order) %>% # Grouped by Order
  # group_by(Family) %>% # Grouped by Family
  # group_by(Genus) %>% # Grouped by Genus
  # group_by(`Curated ID`) %>% # Grouped by Species
  reframe(
    "Total reads" = sum(Reads),
    "ASVs" = sum(ASVs),
    "OTUs" = sum(OTUs),
    "Orders" = length(unique(`Order`)),
    "Families" = length(unique(`Family`)),
    "Genera" = length(unique(`Genus`)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & !grepl("sp\\.", `Curated ID`)])),
    "Nspecies" = length(unique(`Curated ID`)) - Species
  ) %>%
  unique()

# View(grouped_sampled_stats)
View(grouped_filt_stats)

# Mean and SD of alpha-diversity

grouped_filt_site <- grouped_filt_tbl %>% #dependendo da tabela, resultados diferentes
  group_by(Level, Site) %>% 
  mutate(
    "Diversity" = length(unique(`Curated ID`))) %>%
  ungroup() %>% 
  select(Level, Site, Diversity) %>% 
  unique() %>% 
  group_by(Level) %>% 
  mutate(
    "Mean" = mean(Diversity),
    "SD" = sd(Diversity))



```


## Exploratory tables

```{r, eval=FALSE,echo=FALSE}

## Exploratory tables ----

# Reads through data filtering 
raw_results_reads <- raw_results_tbl %>% group_by(Sample) %>% 
  summarise(`Raw reads` = sum(Abundance))

pre_filt_reads <- pre_filt_tbl %>% group_by(Sample) 

filt_reads <- filt_tbl %>% group_by(Sample) %>% 
  summarise(`Filt reads` = sum(Abundance))

evol_reads <- raw_results_reads %>%
  left_join(pre_filt_reads, by = "Sample") %>%
  left_join(filt_reads, by = "Sample") %>% 
  group_by(`Raw reads`, `Filt reads`) %>% 
  select(Sample, `Raw reads`, `Filt reads`) %>% 
  unique()

View(evol_reads)

# ASVs that weren't identified at species Level (genus + epithet)
spp_level <- grouped_filt_tbl %>%
  reframe("Species" = unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & 
                                            !grepl("sp\\.", `Curated ID`)]))
nspp_level <- grouped_filt_tbl %>% 
  reframe("Nspecies" = setdiff(unique(`Curated ID`), spp_level$Species))
    
View(nspp_level) 

# Informations grouped by taxa detected 
id_level <- grouped_filt_tbl %>%
  group_by(`Curated ID`) %>%
    reframe("Order" = Order, "Family" = Family, "Genus" = Genus,
      "Curated identification" = `Curated ID`,
      "Reads" = sum(Reads), 
      "ASVs" = sum(ASVs),
      "OTUs" = sum(OTUs),
    ) %>% 
    unique() %>% 
    select("Order", "Family", "Genus",
           # "Maior hit do BLASTn",
           "Curated identification",
           "Reads", "ASVs","OTUs") 

View(id_level)

# Arquivo para artigo
write.xlsx(id_level, paste0(tbl_path, "/id_level_tbl.xlsx"))

# Informations grouped  by Class
class_Level <- grouped_filt_tbl %>%
  reframe(
    "Total reads" = sum(Reads), 
    "ASVs" = sum(ASVs),
    "OTUs" = sum(OTUs),
    "Orders" = length(unique(Order)),
    "Families" = length(unique(Family)),
    "Genera" = length(unique(Genus)),
    "Identifications" = length(unique(`Curated ID`)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & 
                                               !grepl("sp\\.", `Curated ID`)])), # IDs a Level de spp
    "Nspecies" = length(unique(`Curated ID`)) - Species) # IDs em outros niveis taxonomicos
  
View(class_Level)
  
# Informations grouped by Order
order_level <- grouped_filt_tbl %>%
  filter(Filter %in% "MCE") %>%
  group_by(Order) %>%
  reframe(
    # "Curated ID" = `Curated ID`,
    "Order" = Order,
    "Total reads" = sum(Reads), # dt_all_resume && # dt_filt
    "ASVs" = sum(ASVs),
    "OTUs" = sum(OTUs),
    "Families" = length(unique(Family)),
    "Genera" = length(unique(Genus)),
    "Identifications" = length(unique(`Curated ID`)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & !grepl("sp\\.", `Curated ID`)])), # a nivel de spp
    "Nspecies" = length(unique(`Curated ID`)) - Species # em outros niveis alem de spp
    ) %>% 
  unique()
  
View(order_level)

# Arquivo para artigo
# write.xlsx(order_level, paste0(tbl_path, "/order_level_tbl.xlsx"))
write.xlsx(order_level, paste0(tbl_path, "/order_level_tbl.xlsx"))

# Informations grouped by Family
family_level <- grouped_filt_tbl %>%
  filter(Filter %in% "MCE") %>%
  group_by(Family) %>%
  reframe(
    "Order" = Order,
    "Total reads" = sum(Reads), 
    "ASVs" = sum(ASVs),
    "OTUs" = sum(OTUs),
    "Genera" = length(unique(Genus)),
    "Identifications" = length(unique(`Curated ID`)),
    "Species" = length(unique(`Curated ID`[grepl("^[A-Za-z]+\\s[A-Za-z]+$", `Curated ID`) & !grepl("sp\\.", `Curated ID`)])), # a nivel de spp
    "Nspecies" = length(unique(`Curated ID`)) - Species # em outros niveis alem de spp
    ) %>% 
  unique()
  
View(family_level)

# Informations grouped by Sample
grouped_by_Sample <- filt_tbl %>%
 # filter(Expedition %in% "Novembro 2021") %>%
 select(Sample, Filter, Reads, Site) %>%
 group_by(Filter,
          Sample
          ) %>%
 mutate("Total reads" = sum(Reads)) %>%
 reframe(Filter,
         # Sample,
         Site,
         `Total reads`) %>%
 unique()

# View(grouped_by_Sample)

# Especies encontradas em apenas uma amostra
Counts <- grouped_filt_tbl %>% 
  group_by(`Curated ID`) %>%
  # group_by(`Curated ID`, Filter) %>%
  reframe(count = n_distinct(Sample),
            Filter, Sample) %>% 
  unique()

unique_counts <- 
  Counts %>% filter(count == 1) %>% 
  reframe(`Curated ID`,
          Sample)

View(unique_counts)
```

## Exploratory plots

```{r, eval=FALSE,echo=FALSE}
## Plots ----

# Proportions of identified orders HoH suggestions
fish_ord_all <- grouped_filt_tbl %>% 
  unite(Site, Year, Filter, sep = "_", remove = FALSE, col = "un_amostral") %>% 
  select(c( Order,`Curated ID`, Site, Year, Filter, un_amostral)) %>% 
  unique() %>%
  group_by(Site, Year, Filter) %>% 
  mutate("alpha_d" = length(`Curated ID`)) %>%
  ungroup() %>% 
  group_by(Order,Site, Year, Filter) %>% 
  mutate("alpha_d_order" = length(`Curated ID`)) %>% 
  ungroup() %>% 
  mutate("alpha_d_order_%" = round(alpha_d_order/alpha_d*100, digits = 2)) %>% 
  select(-c(`Curated ID`)) %>% 
  unique() %>% 
  group_by(un_amostral) %>% 
  mutate(Year_Level = case_when(Year == "2020" ~ "High",
                                Year == "2021" ~ "Low",
                                Year == "2022" ~ "High")) %>%
  filter(Filter %in% "MCE") %>% 
  ggplot(aes(
             # x = Site,
             x = Year_Level,
             y = alpha_d_order,
             fill = Order)) +
  geom_col(position = "fill") +
  geom_text(aes(label = sprintf("%0.1f%%",`alpha_d_order_%`)), # adicionar % dentro das barras
            position = position_fill(vjust = 0.5),
            color = "black") +
  # facet_grid(rows = vars(Year_Level),
  facet_grid(rows = vars(Site),
  space = "free", 
  scales = "free",
  drop = TRUE) +
  coord_flip() + #virando 90o o grafico
  guides(fill = guide_legend("Orders")) + ## alterar o titulo da legenda 
  theme(panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2))
        ) +
  theme(plot.title = element_text(size = 20, face = "bold"),
        axis.text.x = element_text(size = 20), ## definindo o tamanho de cada texto
        axis.text.y = element_text(size = 20),
        axis.title.x = element_text(size = 20, face = "bold"),
        axis.title.y = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 19),
        legend.title = element_text(size = 19),
        strip.text = element_text(size = 15, face = "bold"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "lightgray", color = NA)) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("#1f77b4",
                               "#ff7f0e",
                               "#2ca02c",
                               "#d62728"
                               )) +
  labs(x = "Reservoir level", y = "Proportion of species") 

fish_ord_all

ggsave(plot = fish_ord_all, 
       filename = paste0(figs_path, "/alpha_ord_HoH", "-", Sys.Date(), ".pdf", sep = ""),
       units = "cm", 
       height = 20, 
       width = 30, 
       dpi = 600)

# Proportions of identified families 
fish_fam_all <- grouped_filt_tbl %>%
  unite(Site, Year, Filter,sep = "_", remove = FALSE, col = "un_amostral") %>% 
  select(c(Order, Family,`Curated ID`, Site, Year, Filter, un_amostral)) %>% 
  unique() %>%
  group_by(Site, Year, Filter) %>% 
  mutate("alpha_d" = length(`Curated ID`)) %>%
  ungroup() %>% 
  group_by(Family,Site, Year, Filter) %>% 
  mutate("alpha_d_order" = length(`Curated ID`)) %>% 
  ungroup() %>% 
  mutate("alpha_d_order_%" = round(alpha_d_order/alpha_d*100, digits = 2)) %>% 
  select(-c(`Curated ID`)) %>% 
  unique() %>% 
  group_by(un_amostral) %>% 
  mutate(Year_Level = case_when(Year == "2020" ~ "Full",
                                Year == "2021" ~ "Low",
                                Year == "2022" ~ "Full")) %>%
  filter(Filter %in% "MCE") %>% 
  ggplot(aes(x = Year_Level,
             y = alpha_d_order,
             fill = Family)) +
  geom_col(position = "fill") +
  geom_text(aes(label = sprintf("%0.1f%%",`alpha_d_order_%`)), #adicionando % dentro das barras
            position = position_fill(vjust = 0.5),
            color = "black") +
  # facet_grid(rows = vars(Year_Level),
  facet_grid(rows = vars(Site),
  space = "free", 
  scales = "free",
  drop = TRUE) +
  coord_flip() + #virando 90o o grafico
  guides(fill = guide_legend("Family")) + ## alterar o titulo da legenda 
  theme(panel.grid.major = element_line(color = "grey",
                                        linewidth = 0.2,
                                        linetype = 1),
        axis.ticks = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(color = "black", size = rel(1.2)),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black", size = rel(1.2)),
        plot.title = element_text(color = "black", size = rel(1.5)),
        plot.subtitle = element_text(color = "black", size = rel(1.2))
        ) +
  theme(plot.title = element_text(size = 20, face = "bold"),
        axis.text.x = element_text(size = 20), ## definindo o tamanho de cada texto
        axis.text.y = element_text(size = 20),
        axis.title.x = element_text(size = 20, face = "bold"),
        axis.title.y = element_text(size = 20, face = "bold"),
        legend.text = element_text(size = 19),
        legend.title = element_text(size = 19),
        strip.text = element_text(size = 15, face = "bold"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "lightgray", color = NA)) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("#5d9cecff",
                               "#4fc1e9ff",
                               "#48cfadff",
                               "#a0d468ff",
                               "#ffce54ff",
                               "#fc6e51ff",
                               "#ed5565ff",
                               "#ac92ecff",
                               "#ec87c0ff"
                               )) +
  labs(x = "Reservoir level", y = "Proportion of species")

fish_fam_all

ggsave(plot = fish_fam_all, 
         filename = paste0(figs_path, "/alpha_fam_HoH", "-", Sys.Date(), ".pdf", sep = ""),
       units = "cm", 
       height = 20.5, 
       width = 30, 
       dpi = 600)

# Tile plots 
{
  # Defining levels 
  {
    # Defining order levels
    order_levels <- grouped_filt_tbl$Order %>% 
      sort() %>% unique()
    
    # Defining spp levels
    spp_levels <- grouped_filt_tbl$`Curated ID` %>% 
      sort() %>% unique()
  }
  
  # Tile plots 
  {
    # All Years (only MCE Filters) 
    tile_all <- grouped_filt_tbl %>%
      # filter(Level %in% "Full") %>% 
      # filter(Level %in% "Low") %>% 
      mutate(`Curated ID` = factor(`Curated ID`, levels = rev(spp_levels)),
             Site = factor(Site),
             Year_Level = case_when(Year == "2020" ~ "High",
                                    Year == "2021" ~ "Low",
                                    Year == "2022" ~ "High"),
             Order_abrev = case_when(Order == "Characiformes" ~ "Char",
                                     Order == "Cichliformes" ~ "Cich",
                                     Order == "Siluriformes" ~ "Sil",
                                     Order == "Cyprinodontiformes" ~ "Cypr"),
             Order_abrev = factor(Order_abrev, levels = c("Char", "Cich", 
                                                          "Cypr", "Sil"))) %>%
      ggplot(aes(
        y = `Curated ID`,
        # x = Year_Level,
        x = Site,
        fill = RRA
        )) +
      geom_tile() +
      #geom_text(aes(label= sprintf("%0.2f", round(RRA, digits = 2))), 
      # stat = "identity",
      # colour = "black", size = 4) +
      facet_grid(
        # cols = vars(Site),
        cols = vars(Year_Level),
        rows = vars(Order_abrev),
        # rows = vars(Order),
        space = "free",
        scales = "free",
        drop = TRUE
        ) +
      scale_fill_gradientn(
        name = "RRA (%)",
        colours = rev(c("#30123BFF", "#4662D7FF", "darkgreen", "#72FE5EFF", "#C7EF34FF", "#FABA39FF", "#F66B19FF", "#CB2A04FF", "#7A0403FF")),
        values = scales::rescale(c(0, 0.01, 0.05, 0.25, 1, 2.5, 5, 10, 25, 50, 75, 100)),
        breaks = c(0, 0.01, 0.05, 0.25, 1.00, 2.50, 5.00, 10.00, 25.00, 50.00, 100.00),
        labels = scales::number_format(accuracy = 0.01),
        limits = c(0.01, 100),
        na.value = "#7A0403FF",
        trans = "log10"
        ) +
      theme(
        # panel.background = element_blank(),
        panel.grid.major = element_line(color = "grey", linewidth = 0.2, linetype = 1),
        legend.text = element_text(size = 13, color = "black"),
        legend.title = element_text(size = 14, face = "bold", margin = margin(b = 10), color = "black"),
        strip.background = element_rect(fill = "#e4e4e4"),
        strip.text = element_text(size = 17, face = "bold", color = "black"),
        axis.text.x = element_text(size = 15),
        axis.text.y = element_text(size = 17, face = "italic"),
        axis.title = element_text(size = 20, face = "bold"),
        legend.position = "right",
        legend.key.height = unit(3, 'cm')
        ) +
      labs(
        fill = 'RRA (%)',
        x = "Sampling site",
        y = "Species"
        )
    
    tile_all
    
    ggsave(plot = tile_all,
           filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
                            "tile_20-22_MCE", "-", Sys.Date(), ".pdf", sep = ""),
           units = "cm", 
           height = 20.5, 
           width = 25.5, 
           dpi = 600)
  }
}

# Alpha diversity 

## Tax table
# blast_tax_tbl <- grouped_sampled_filt_tbl %>% 
blast_tax_tbl <- grouped_filt_tbl %>% 
  dplyr::select(c("Curated ID",
                  "Genus",
                  "Family",
                  "Order",
                  "Class")
                ) %>% 
  unique() %>% 
  as.data.frame() %>% 
  `rownames<-`(.$`Curated ID`) %>% 
  dplyr::select(-c("Curated ID")) %>% 
  as.matrix()
  
## OTU table

# otu_tbl <- grouped_sampled_filt_tbl %>%
otu_tbl <- grouped_filt_tbl %>%
  unite(Site, Level, Year, col = "Sample", remove = FALSE) %>%
  # select(Sample, RRA, `Curated ID`) %>% 
  # select(Sample, Abundance, `Curated ID`) %>% 
  # select(Abundance, `Curated ID`) %>% 
  select(Sample, Reads, `Curated ID`) %>%
  pivot_wider(
    id_cols = `Curated ID`,
    # values_from = RRA,
    # values_from = Abundance,
    values_from = Reads,
    values_fn = sum_uniq,
    names_from = Sample,
    names_sort = TRUE) %>% 
  mutate(across(starts_with("P") ,~ replace_na(.,replace = 0))) %>%
  # mutate(across(starts_with("L") ,~ replace_na(.,replace = 0))) %>%
  mutate_if(is.numeric, jaccarize) %>% 
  as.data.frame() %>% 
  column_to_rownames(var = "Curated ID") %>% 
  dplyr::filter(rowSums(.)!= 0) %>%
  dplyr::select(which(!colSums(., na.rm = TRUE) %in% 0))

## Metadata table

metadata_tbl <-
  # grouped_sampled_filt_tbl %>%
  grouped_filt_tbl %>%
  ungroup() %>%
  unite(Site, Level, Year, col = "Sample", remove = FALSE) %>%
  select(Sample, Site, Level, Year) %>%
  as.data.frame() %>% 
  unique() %>% 
  `rownames<-`(.$Sample)

## Creating Phyloseq object

phyloseq_tbl <- phyloseq(otu_table(otu_tbl, ,taxa_are_rows = TRUE),
                         sample_data(metadata_tbl),
                         tax_table(blast_tax_tbl)) 
## Estimating richness

rich_tbl <- estimate_richness(phyloseq_tbl,
                              split = TRUE,
                              measures = c("Observed", "Chao1","Obs./Chao1",
                                           "se.chao1","ACE", "se.ACE",
                                           "Shannon","Simpson", "InvSimpson")) %>% 
  as_tibble(rownames = "Sample") %>% 
  mutate("Obs./Chao1" = Observed/Chao1) %>% 
  pivot_longer(cols = c("Observed", "Chao1","Obs./Chao1", "se.chao1",
                        "ACE", "se.ACE", "Shannon", "Simpson", "InvSimpson"),
               names_to = "Indice",
               values_to = "Values") %>%
  mutate(Indice = factor(Indice, levels = c("Observed", "Chao1",
                                            "Obs./Chao1", "se.chao1",
                                            "ACE", "se.ACE", "Shannon",
                                            "Simpson", "InvSimpson"))) %>% 
  left_join(y = metadata_tbl,
            by = "Sample")
  
boxplot_alpha <- rich_tbl %>% 
  filter(Indice %in% "Observed") %>% 
  ggplot(aes(x = Level,
             y = Values)) +
  geom_boxplot(fill = c("#3381b1", "#f1c232"),
               alpha = 0.7) +
  geom_jitter(width = 0.2, alpha = 0.5) +
  theme(panel.grid.major = element_line(color = "grey", linewidth = 0.2, linetype = 1),
        strip.background = element_rect(fill = "#e4e4e4"),
        strip.text = element_text(size = 17, face = "bold", color = "black"),
        axis.text.x = element_text(size = 17),
        axis.text.y = element_text(size = 17),
        axis.title = element_text(size = 20, face = "bold")) +
      labs(x = "Water level",
           y = "Richness")

boxplot_alpha

ggsave(plot = boxplot_alpha,
           filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
                            "boxplot_alpha_levels", "-", Sys.Date(), ".pdf", sep = ""),
           units = "cm", 
           height = 9, 
           width = 10, 
           dpi = 600)
  

# Venn diagram 
{
  # Filtering
  resume_venn <- grouped_filt_tbl %>%
    select(`Curated ID`, Genus, Family, Order, Filter, Level, ) %>%
    unique()
  
  # Common spp between High and Low
  spp_comm <- resume_venn %>% 
    filter(Filter %in% "MCE") %>%
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(n_Level = n_distinct(Level)) %>%
    filter(n_Level >= 2) %>%
    select(`Curated ID`) # Trocar para spp, genero, familia e ordem
  View(spp_comm_2)
  
  # Exclusive sppHigh and Low
  spp_diff <- resume_venn %>%
    filter(Filter %in% "MCE") %>%
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(Low_count = sum(as.integer(Level == "Low")),
              High_count = sum(as.integer(Level == "Full")),
              Low_only = Low_count >= 1 & High_count == 0,
              High_only = Low_count == 0 & High_count >= 1) %>% 
    ungroup()
  View(spp_diff)
  
    # Common spp between Sample sites
  spp_comm_2 <- resume_venn %>% 
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(n_Level = n_distinct(Site)) %>%
    filter(n_Level == 4) %>%
    select(`Curated ID`) # Trocar para spp, genero, familia e ordem
  View(spp_comm_3)
  
  # Exclusive spp Sample sites
  spp_diff_2 <- resume_venn %>%
    group_by(`Curated ID`) %>% # Trocar para spp, genero, familia e ordem
    summarise(P3_count = sum(as.integer(Site == "P3")),
              P2_count = sum(as.integer(Site == "P2")),
              P4_count = sum(as.integer(Site == "P4")),
              P1_count = sum(as.integer(Site == "P1")),
              P3_only = P3_count >= 1 & P2_count == 0 & P4_count == 0 & P1_count == 0,
              P2_only = P2_count >= 1 & P3_count == 0 & P4_count == 0 & P1_count == 0,
              P4_only = P4_count >= 1 & P3_count == 0 & P2_count == 0 & P1_count == 0,
              P1_only = P1_count >= 1 & P3_count == 0 & P2_count == 0 & P4_count == 0) %>% 
    ungroup()
  View(spp_diff_2)
}

```

## Analysis 

```{r, eval=FALSE,echo=FALSE}
## PCoA ----

# Dados

## Gerando uma nova tabela wider_filt_tbl_redux com amostras artificiais
{
wider_filt_tbl_redux <- grouped_filt_tbl %>% 
  mutate(RRA = round(RRA,digits = 4)) %>% 
  ungroup() %>% 
  unite(Site, Level, Year, col= "New_sample", remove = FALSE)

# Converter planilha de identificações por ponto para o formato Amostras X IDs
{
  # Função necessária para juntar as abds da tabela em IDs
  sum_uniq <- function(vec) {
    if (is.character(vec)) {
      suniq <- BiocGenerics::unique(vec)
    }
    if (is.numeric(vec)) {
      suniq <- sum(vec)
    }
    return(suniq)
  }
  
  # Número de linhas (IDs diferentes) por amostra
  wider_filt_tbl_redux$New_sample %>% table()
  
  wider_filt_tbl_redux %>% colnames()
  
  FINAL_tbl_IDs_2 <- wider_filt_tbl_redux %>%
    select(c("New_sample", "Curated ID", "Year", "Level", "Site", "Filter", "RRA")) %>% 
    pivot_wider(id_cols = c("New_sample", "Year", "Level", "Site", "Filter"),
                names_from = `Curated ID`,
                values_from = `RRA`,
                values_fn = sum_uniq,
                names_sort = TRUE,
                names_prefix = "ID_") %>% 
    relocate(c("New_sample", "Year", "Level", "Filter", "Site", starts_with("ID_"))) %>%  
    mutate(across(starts_with("ID_"), \(x) replace_na(x, 0))) %>% 
    mutate("Site" = as.factor(Site)) %>% 
    mutate("Level" = as.factor(Level)) 
  
  # Verificando a soma das linhas
  FINAL_tbl_IDs_2 %>% select(starts_with(match = "ID_")) %>% rowSums(na.rm = TRUE)
}

# Preparar dados para entrada no pacote vegan
{
  colnames(FINAL_tbl_IDs_2)
  
  FINAL_tbl_IDs_2$`New_sample` %>% unique() %>% sort()
  
  all_IDs_PCoA_tbl <- FINAL_tbl_IDs_2 %>% 
    mutate("Sample number" = 0) %>% 
    relocate("Sample number" )
  
  # Associar números de amostra aos nomes de amostra
  for (sample in 1:nrow(all_IDs_PCoA_tbl)) {
    all_IDs_PCoA_tbl$`Sample number`[sample] <- sample
  }
  
  # Ordenar dataframe usado no NMDS
  all_IDs_PCoA_df <- all_IDs_PCoA_tbl %>%
    # mutate(Level = ifelse(Level == "Full", "Full", "Empty")) %>%
    as.data.frame()
  
  # Nomear linhas como números de amostra
  row.names(all_IDs_PCoA_df) <- all_IDs_PCoA_df$`Sample number`
  all_IDs_PCoA_df <- all_IDs_PCoA_df 
  
  # Corrigir nomes das espécies para evitar problemas na plotagem
  colnames(all_IDs_PCoA_df)
  colnames(all_IDs_PCoA_df)[8:ncol(all_IDs_PCoA_df)] <- colnames(all_IDs_PCoA_df)[8:ncol(all_IDs_PCoA_df)] %>%
    str_replace_all(pattern = " ", replacement = "_") %>% 
    str_replace_all(pattern = "\\.", replacement = "") %>% 
    str_replace_all(pattern = "\\(", replacement = "") %>% 
    str_replace_all(pattern = "\\)", replacement = "")
}
}

## Executando o PCoA para High versus Low
{
  # Filtrando os dados

MCE_all_IDs_PCoA_df <- all_IDs_PCoA_df %>% 
  select(where(~ any(. != 0))) 

colnames(MCE_all_IDs_PCoA_df[,8:ncol(MCE_all_IDs_PCoA_df)])

#Criando a Matriz de distancia
pcOa_dist <- vegan::vegdist(x = MCE_all_IDs_PCoA_df[,8:ncol(MCE_all_IDs_PCoA_df)],
                            method = "jaccard",
                            binary = TRUE)

pcOa <- cmdscale(pcOa_dist, eig = TRUE)

ordiplot(pcOa, display = 'sites', type = 'text')

# Porcentagens

percent_pcOa <- pcOa$eig / sum(pcOa$eig) * 100
percent_pcOa[1:2]

# Espécies 

# Fazer o fit das espécies para identificar a significância delas na explicação dos agrupamentos. 
# Esse é o passo que mais demora quando com muitas amostras e espécies.
meta.spp.fit <- envfit(pcOa, 
                       MCE_all_IDs_PCoA_df[,8:ncol(MCE_all_IDs_PCoA_df)], 
                       permutations = 999) # this fits species vectors

# Obter valores de p para as espécies
sps_pvals <- tibble("IDs" = names(meta.spp.fit$vectors$pvals),
                    "p-value" = meta.spp.fit$vectors$pvals)

spp.scrs <- as.data.frame(scores(meta.spp.fit, display = "vectors")) %>%
  mutate("IDs" = rownames(.)) %>%
  left_join(y = sps_pvals, by = "IDs")

{
  spp.scrs$IDs <- gsub("ID_", "", spp.scrs$IDs)
  spp.scrs$IDs <- gsub("_", " ", spp.scrs$IDs)
  } 

# Selecionar espécies significativas
sig.spp.scrs <- spp.scrs 

t.spp.scrs <- spp.scrs %>%
  filter(`p-value` <=
           0.05) # definir p-value aqui!

# Pontos amostrais

#Definir os valores de PCoA1 e PCoA2, e os metadados de cada amostra/ponto amostral
site.scrs <- as.data.frame(scores(pcOa, display = "sites"))

colnames(site.scrs) <- c("PCoA1", "PCoA2")

site.scrs <- site.scrs %>% 
  mutate("Sample number" = as.double(row.names(.))) %>%
  left_join(y = MCE_all_IDs_PCoA_df[, c("Sample number",
                                    "New_sample",
                                    "Level",
                                    "Site"
  )],
  by = "Sample number")

# Determinar centroides
scrs <- scores(pcOa, display = "sites")

cent <- aggregate(scrs ~ Level, data = site.scrs, FUN = "mean")

# Calcular elipses
PCoA <- data.frame("PCoA1" = pcOa$points[, 1],
                   "PCoA2" = pcOa$points[, 2],
                   # "Filter" = as.factor(MCE_all_IDs_PCoA_df$Filter), check.names = FALSE)
                   "Level" = as.factor(MCE_all_IDs_PCoA_df$Level), check.names = FALSE)

PCoA.mean <- aggregate(PCoA[, 1:2], list(group = PCoA$Level), "mean")

# Elipses

# plot(all_ps_vegan_ord_meta)
ordiplot(pcOa, display = 'sites', type = 'text')

# Sobrepor as elipses
ord <- ordiellipse(ord = pcOa, 
                   groups = MCE_all_IDs_PCoA_df$Level,
                   display = "sites",
                   kind = "ehull", conf = 0.95, label = T)


# Funcao do vegan de calcular elipses
veganCovEllipse <-function (cov, center = c(0, 0), scale = 1, npoints = 100) 
{
  theta <- (0:npoints) * 2 * pi/npoints
  Circle <- cbind(cos(theta), sin(theta))
  t(center + scale * t(Circle %*% chol(cov)))
}

df_ell <- data.frame()

for(g in levels(PCoA$Level)){
  print(g)
  df_ell <- 
    rbind(df_ell, 
          cbind(as.data.frame(with(PCoA[PCoA$Level==g,],
                                   veganCovEllipse(
                                     ord[[g]]$cov,
                                     ord[[g]]$center,
                                     ord[[g]]$scale))),
                Level=g))}

# Configurar plot PCoA

# Definir paleta de cores

cores <- c("#3381b1",
           #"#f1c232",
           "#E63946")

# Plot

  PCoA_MCE_2 <- 
    ggplot(data = site.scrs,
           aes(x=PCoA1, 
               y=PCoA2)) +
    # Elipses 
    ggforce::geom_mark_ellipse(inherit.aes = FALSE,
                               data = df_ell,
                               aes(x = Dim1,
                                   y = Dim2,
                                   group = Level,
                                   label = Level,
                                   col = Level,
                                   fill = Level),
                               alpha=0.10,
                               linetype=2,
                               expand = 0,
                               label.fontsize = 20,
                               con.cap = 0.1) +
    # Niveis amostrais
    geom_point(aes(x=PCoA1,
                   y=PCoA2,
                   fill = Level,
                   # label = `Sampling unit`,  #descomentar se quiser exibir os nomes dos `Sampling sites`s
                   col=Level,
                   group = Level,
                   shape = Site),
               stroke = 0.5,
               alpha = 0.75,
               size = 3) +
    # Nomes dos Sampling sites amostrais
    ggrepel::geom_text_repel(aes(label = Site),  #descomentar bloco se quiser exibir os nomes dos pontos
                             # hjust=0.5,
                             # vjust=2.75,
                             size=6,
                             direction = "both",
                             segment.size = 0.25,
                             segment.alpha=0.1,
                             # min.segment.length = 1.5,
                             force = 3,
                             max.overlaps = 100,
                             fontface = "bold") +
    # Vetores das IDs
    geom_segment(data = sig.spp.scrs, aes(x = 0,
                                          xend = Dim1,
                                          y = 0,
                                          yend = Dim2),
                 arrow = arrow(length = unit(0.1, "cm")),
                 colour = "grey30",
                 alpha = 0.5,
                 lwd = 0.3) + #add vector arrows of significant species
    # Nomes das IDs
    geom_text_repel(data = sig.spp.scrs,
                    aes(x=Dim1, y=Dim2, label = IDs),
                    size = 3.5,
                    alpha = 0.75,
                    # direction = "both",
                    segment.size = 0.25,
                    segment.alpha = 0.1,
                    max.overlaps = 25) +
    # Centroides
    geom_point(data = cent,
               aes(x = Dim1,
                   y = Dim2, # colour = Filter,
                   fill = Level),
               size = 8,
               colour = "#222222",
               alpha = 0.75,
               shape  = 13) + 
    coord_fixed(expand = c(0.5))+
    theme(panel.grid.major = element_line(color = "grey",
                                          size = 0.2,
                                          linetype = 1),
          axis.text = element_text(size = 17, color = "black"),
          axis.title = element_text(size = 20, face = "bold"),legend.position = "none") +
    labs(x = "PCoA1 (28.76%)",
         y = "PCoA2 (23.87%)") +
    scale_fill_manual(values = cores) +
    scale_colour_manual(values = cores)
  
  PCoA_MCE_2
  
  # Salvar em pdf
  ggsave(plot = PCoA_MCE_2,
         filename = paste("/home/gabriel/projetos/LI_paper/results/figures/",
                          "PCoA_MCE_all",
                          "-", Sys.Date(), ".pdf", sep = ""),
         units = "cm",
         height = 16.75,
         width = 16.75,
         dpi = 600)
}
}

## Permanova
{
  ## Primeiro e necessario gerar uma matriz de distancias para plotar as amostras
  ## e em seguida testar a assumpcao de que os dados dispersos formam grupos bem definidos.
  
  # Distance matrix
  
  pcOa_mtx <-  as.data.frame(as.matrix(pcOa_dist)) # gerado anteriormente para realizar o plot PCoA
  
  #Assumptions
  
  Dispersion <- betadisper(pcOa_dist, group = 
                             # mce_stx_IDs_NMDS_df$Filter, 
                             MCE_all_IDs_PCoA_df$Level, 
                           type = "centroid") # testando a dispercao
  
  plot(Dispersion)
  
  anova(Dispersion)
  
  #Test
  
  perma_test <- adonis2(pcOa_mtx ~ as.factor(MCE_all_IDs_PCoA_df$Level), data = pcOa_mtx,
                   permutations=9999)
  perma_test
}  

## Species accumulation curves ----

# All samples (Richness vs Samples)
{
  ## Definindo os dados 
  wider_grouped_sampled_filt <- grouped_sampled_filt_tbl %>%
    filter(Filter %in% "MCE") %>% 
    select(c(Sample,
             Site,
             Level,
             `Curated ID`,
             Reads,
             Expedition,
             Filter,
             Level)) %>%
    group_by(Sample, `Curated ID`, Reads) %>%
    unique %>%
    pivot_wider(names_from = `Curated ID`,
                values_from = Reads,
                names_sort = TRUE) %>% 
    as.data.frame() %>% 
    mutate(across(everything(), ~replace_na(.x, 0))) %>%
    column_to_rownames("Sample")
  
  grouped_sampled_filt_tbl %>% 
    group_by(Sample, Site, Year, Level, Expedition) %>% 
    select(Sample, Site, Year, Level, Expedition) %>% 
    unique() %>% 
    View() 
  
    ## Function to transform values into 1 and zeros
  jaccarize <- function(x) {
    # Using ifelse to check each element of the vector x
    ifelse(x == 0, 0, 1)
    }

  wider_grouped_sampled_filt <- wider_grouped_sampled_filt %>%
    mutate(across(c(5:ncol(wider_grouped_sampled_filt)), jaccarize))
  
  ## All MCE samples
  accum_mce <- specaccum(wider_grouped_sampled_filt[5:ncol(wider_grouped_sampled_filt)],
                             # method = "rarefaction"
                             method = "random" # same choice as Di Muri et al. 2020
                             # method = "exact"
                             )
  plot(accum_mce$sites, accum_mce$richness)
  
  ## Getting into a dataframe
  accum_mce_df <- tibble("Sites" = c(0, accum_mce$sites),
                         "Individuals" = c(0, accum_mce$individuals),
                         "Richness" = c(0, accum_mce$richness),
                         "sd" = c(0, accum_mce$sd),
                         "Level" = "All")

  ## Low-level samples only
  wider_grouped_sampled_filt_low <- wider_grouped_sampled_filt %>%
    filter(Level %in% "Low")

  accum_low <- specaccum(wider_grouped_sampled_filt_low[5:ncol(wider_grouped_sampled_filt_low)],
                         # method = "rarefaction"
                         method = "random" # same choice as Di Muri et al. 2020
                         # method = "exact"
                         )
  
  ## Getting into a dataframe
  accum_low_df <- tibble("Sites" = c(0, accum_low$sites),
                         "Individuals" = c(0, accum_low$individuals),
                         "Richness" = c(0, accum_low$richness),
                         "sd" = c(0, accum_low$sd),
                         "Level" = "Low")
  ## High-level samples only
  wider_grouped_sampled_filt_high <- wider_grouped_sampled_filt %>%
    filter(Level %in% "High")

  accum_full <- specaccum(wider_grouped_sampled_filt_high[5:ncol(wider_grouped_sampled_filt_high)],
                         # method = "rarefaction"
                         method = "random" # same choice as Di Muri et al. 2020
                         # method = "exact"
                         )
  
  ## Getting into a dataframe
  accum_full_df <- tibble("Sites" = c(0, accum_full$sites),
                          "Individuals" = c(0, accum_full$individuals),
                          "Richness" = c(0, accum_full$richness),
                          "sd" = c(0, accum_full$sd),
                          "Level" = "High")
  
  ## Combining dataframes
  combined_accum <- bind_rows(accum_mce_df,
                              accum_low_df,
                              accum_full_df
                              )
  
  combined_accum$Level <- factor(combined_accum$Level,
                                  levels = c("All",
                                             "Low",
                                             "High"
                                             ))
  # Definir paleta de cores
  
  cores <- c("#3381b1",
             "#f1c232")
  
  ## Plotting
  collector_plot_all <-
    combined_accum %>% 
    ggplot(aes(x = Sites,
               y = Richness,
               color = Level,
               fill = Level )) +
    geom_line(linewidth = 2, alpha = 0.50) +
    geom_point(size = 1) +
    # geom_ribbon(aes(ymin = Richness - sd, ymax = Richness + sd),
    #             alpha = 0.15, size = 0.1) +
    # geom_linerange(aes(ymin = Richness - sd, ymax = Richness + sd),
               # size = 0.3, alpha = 0.6) +
    # geom_line(aes(y = Richness + sd, color = Level), linetype = "dashed", alpha = 0.5) +
    # geom_line(aes(y = Richness - sd, color = Level), linetype = "dashed", alpha = 0.5) +
    # geom_pointrange(aes(ymin = Richness - sd, ymax = Richness + sd, color = Level), 
    #             alpha = 1) +
    geom_errorbar(aes(ymin = Richness - sd, ymax = Richness + sd, colour=Level), width=.2
                  # ,position= position_dodge(0.1)
                  ) +
    # scale_color_manual(values = c("#2A6B33",  "#3381b1", "#F5AF0F"),
    scale_color_manual(values = c("#2A6B33", "#E63946", "#457B9D"),
                       labels = c("All" = "All samples", 
                                  "Low" = "Low-level", 
                                  "High" = "High-level")) +
    scale_fill_manual(values = c("#3A9647", "#fd2c3b", "#3381b1")) +
    # scale_fill_manual(values = c("#1b9e77", "#d95f02", "#7570b3")) +
    scale_x_continuous(breaks = seq(0, max(combined_accum$Sites), by = 3)) +
    scale_y_continuous(breaks = seq(0, max(combined_accum$Richness), by = 3)) +
    geom_label_repel(data = combined_accum %>%
                       arrange(desc(Sites)) %>%
                       distinct(Level, .keep_all = TRUE),
                     aes(label = recode(Level,
                                        "All" = "All samples",
                                        "Low" = "Low-level",
                                        "High" = "High-level"),
                         color = Level),
                     size = 7,
                     fontface = "bold",
                     nudge_y = 3,
                     fill = "white",
                     box.padding = 1,
                     point.padding = 0.5,
                     segment.color = "grey50",
                     show.legend = FALSE) +
    theme(panel.grid.major = element_line(color = "grey", linewidth = 0.2, linetype = 1),
          axis.text = element_text(color = "black", size = 17),
          axis.title = element_text(color = "black", size = 20, face = "bold"),
          strip.background = element_rect(fill = "#e4e4e4"),
          legend.position = "none") +
    labs(x = "Number of samples",
         y = "Richness")
  
  collector_plot_all
  
  ggsave(file = paste0(figs_path,"/", "collector_plot_all-", Sys.Date(), ".pdf", collapse = ""),
         plot = collector_plot_all,
         width = 20,
         height = 12,
         units = "cm",
         dpi = 300)
    }

## Diversity indexes ----

grouped_sampled_filt_tbl %>% colnames()

# Creating "OTU" table 
{
  gsf_phyloseq <- grouped_sampled_filt_tbl %>%
  pivot_wider(id_cols = c("Curated ID"),
              values_from = RRA,
              values_fn = sum_uniq,
              names_from = "Sample",
              names_sort = TRUE
              ) %>%
  mutate(across(everything(), replace_na, replace = 0)) %>% 
  # mutate_if(is.numeric, round) %>%
  mutate_if(is.numeric, jaccarize) %>%
  as.data.frame() %>%
  `rownames<-`(.$`Curated ID`) %>% 
  select(-c("Curated ID")) %>% 
  filter(rowSums(.) != 0) %>% # Remover linhas cuja soma é zero
  select(which(!colSums(., na.rm=TRUE) %in% 0)) # Remover colunas cuja soma é zero
 
dim(gsf_phyloseq)

gsf_phyloseq %>% rowSums() ==0
}
# Creating taxonomy table
{
grouped_sampled_filt_tbl %>% colnames()

blast_tax_table <- grouped_sampled_filt_tbl %>%
  select(c("Curated ID",
           "Genus",
           "Family",
           "Order"
           )) %>% 
  unique() %>% 
  as.data.frame() %>% 
  `rownames<-`(.$`Curated ID`) %>% 
  select(-c("Curated ID")) %>% 
  as.matrix()
}
# Creating sample metadata table
{
metadata_tbl <- grouped_sampled_filt_tbl %>%
  select(c("Sample", "Site", "Year", "Expedition", "Level", "Filter")) %>% 
  unique() %>%
  as.data.frame() %>% 
  `rownames<-`(.$`Sample`)  
}
# Creating final phyloseq object
{
 final_ps <- phyloseq::phyloseq(otu_table(gsf_phyloseq, taxa_are_rows = T),
                               sample_names(metadata_tbl),
                               tax_table(blast_tax_table))

# Estimating richness (final_ps)

div_est_phylo <- estimate_richness(final_ps,
                                   split = T,
                                   measures = c("Observed", "Chao1", "Obs./Chao1", "se.chao1", 
                                                "ACE", "se.ACE", "Shannon","Simpson", "InvSimpson")) %>%
  as_tibble(rownames = "Sample") %>% 
  mutate("Obs./Chao1" = Observed/Chao1) %>%
  pivot_longer(cols = c("Observed", "Chao1", "Obs./Chao1", "se.chao1",
                        "ACE", "se.ACE", "Shannon","Simpson", "InvSimpson"),
               names_to = "Indice",
               values_to = "Values") %>%
  mutate(Indice = factor(Indice, levels = c("Observed", "Chao1", "Obs./Chao1", "se.chao1",
                                            "ACE", "se.ACE", "Shannon","Simpson", "InvSimpson"))) %>% 
  left_join(y = metadata_tbl,
            by = "Sample")

 ## Sterivex versus MCE (All samples)
 div_stx_mce_plot <-
   div_est_phylo %>%
     # filter(Expedition %in% c("Novembro 2021")) %>%
   filter(Indice %in% c("Observed",
                        "Chao1",
                        "Obs./Chao1",
                        "Shannon",
                        "Simpson")) %>%
   ggplot(aes(y = Values,
              x = Filter ,
              fill = Filter,
              alpha = 0.1)) +
   geom_boxplot(col = "#282828",
                alpha = 1,
                linewidth = 0.3,
                outlier.shape = NA) +
   geom_jitter(aes(col = Site,
                   shape = Level),
               alpha = 0.75,
               height = 0,
               width = 0.25) +
   facet_wrap(nrow = 1,
              ~Indice,
              scales = "free") +
   theme_bw(base_size = 14) +
   theme(axis.text.x = element_text(angle = 45,hjust = 1,vjust = 1),
         axis.title = element_text(size = 14),
         strip.text = element_text(size = 14)) +
   scale_fill_manual(values = c("#f8e199", "#99c0d8")) +
   scale_shape_manual(values = c("Full" = 16, "Low" = 4)) +
   guides(fill = "none") +
   labs(x = "Filter type",
        y = "Value")
 
 div_stx_mce_plot
 
 ## MCE (2020, 2021, 2022)
 div_all_plot <-
   div_est_phylo %>%
     # filter(Filter %in% c("MCE")) %>%
   filter(Indice %in% c("Observed",
                          "Chao1",
                          "Obs./Chao1",
                          "Shannon",
                          "Simpson")) %>%
   ggplot(aes(y = Values,
              x = Level ,
              groups = Level,
              fill = Level,
              alpha = 0.1)) +
   geom_boxplot(col = "#282828",
                alpha = 1,
                linewidth = 0.3,
                outlier.shape = NA) +
   geom_jitter(aes(col = Site,
                   shape = Filter),
               alpha = 1,
               height = 0,
               width = 0.25) +
   # geom_signif(
   #   comparisons = list(c("Full", "Low")),
   #   map_signif_level = F,
   #   col = "#111111",
   #   test = "wilcox.test",
   #   test.args = list(exact = FALSE)
   #   ) +
   facet_wrap(nrow = 1,
              ~Indice,
              scales = "free") +
   theme_bw(base_size = 14) +
   theme(axis.text.x = element_text(angle = 45,hjust = 1,vjust = 1),
         axis.title = element_text(size = 14),
         strip.text = element_text(size = 14)) +
   scale_fill_manual(values = c("#99c0d8","#f8e199")) +
   scale_shape_manual(values = c(1, 3)) +
   guides(fill = "none") +
   xlab(label = "Level") +
   ylab(label = "Value") 

 div_all_plot
 
 # Combining diversity plots in one figure
 
 ggarrange(div_stx_mce_plot, div_all_plot,
                    labels = c("A", "B"),
                    ncol = 2, nrow = 1) 
}

## Diversity indices ----

# Creating "OTU" table

discret_tbl_ID <- grouped_filt_tbl %>%
  select(c(Site,
           `Period abundance`,
           `Curated ID`)) %>%
  mutate(Site_Level = paste(Site, Level, sep = "_")) %>% 
  pivot_wider(id_cols = c("Curated ID"),
              values_from = "Period abundance",
              values_fn = sum_uniq,
              names_from = Site_Level,
              names_sort = TRUE) %>%
  mutate(across(starts_with("SP") ,~ replace_na(.,replace = 0))) %>%
  mutate_if(is.numeric, jaccarize) %>%
  as.data.frame() %>%
  column_to_rownames(var = "Curated ID") %>%
    filter(rowSums(.)!= 0) %>%   #Remover linhas cuja soma é zero
    select(which(!colSums(., na.rm = TRUE) %in% 0))

metadata_tbl_ID <- grouped_filt_tbl %>%
  select(c(Level,
           Site)) %>%
  unique() %>%
  mutate(Site_Level = paste(Site, Level, sep = "_")) %>% 
  as.data.frame() %>%
  `rownames<-`(.$`Site_Level`)
 
# Creating taxonomy table

blast_tax_tbl_ID <- grouped_filt_tbl %>% 
  select(c("Curated ID",
           "Genus",
           "Family",
           "Order",
           "Class")) %>% 
  unique() %>% 
  as.data.frame() %>% 
  `rownames<-`(.$`Curated ID`) %>% 
  select(-c("Curated ID")) %>% 
  as.matrix()

# Creating Phyloseq object

PS_ID <- phyloseq(otu_table(discret_tbl_ID, taxa_are_rows = T),
                  sample_data(metadata_tbl_ID),
                  tax_table(blast_tax_tbl_ID))

div_est_ps_ID <- estimate_richness(PS_ID,
                                   split = TRUE,
                                   c("Observed", "Chao1","Obs./Chao1",
                                     "se.chao1","ACE", "se.ACE",
                                     "Shannon","Simpson", "InvSimpson")) %>% 
  as_tibble(rownames = "Site_Level") %>% 
  mutate("Obs./Chao1" = Observed/Chao1) %>%
  pivot_longer(cols = c("Observed", "Chao1","Obs./Chao1", "se.chao1",
                        "ACE", "se.ACE", "Shannon", "Simpson", "InvSimpson"),
               names_to = "Indice",
               values_to = "Values") %>%
  mutate(Indice = factor(Indice, levels = c("Observed", "Chao1","Obs./Chao1", "se.chao1",
                                            "ACE", "se.ACE", "Shannon", "Simpson", "InvSimpson"))) %>%
  left_join(y = metadata_tbl_ID,
            by = "Site_Level")

# Plotting diversity indices

div_est_ps_ID %>% 
  filter(Indice %in% c("Observed",
                       "Chao1",
                       "Shannon")) %>% 
  ggplot(aes(y = Values,
             x = Site,
             group = Level,
             fill = Level, alpha = 0.1)) +
  geom_col(col = "#282828",
           # alpha = 0.4,
           linewidth = 0.05,
           position = "dodge") +
  # position_dodge() +
  # facet_wrap(nrow = 3,
  #            ~Indice,
  #            scales = "free") +
  # facet_grid(rows = vars(Indice),
  facet_grid(cols = vars(Indice),
             # rows = vars(Level),
             space = "free",
             scales = "free",
             drop = TRUE)

```
